{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene</th>\n",
       "      <th>Glyma.01G000100</th>\n",
       "      <th>Glyma.01G000137</th>\n",
       "      <th>Glyma.01G000174</th>\n",
       "      <th>Glyma.01G000211</th>\n",
       "      <th>Glyma.01G000248</th>\n",
       "      <th>Glyma.01G000285</th>\n",
       "      <th>Glyma.01G000322</th>\n",
       "      <th>Glyma.01G000359</th>\n",
       "      <th>Glyma.01G000400</th>\n",
       "      <th>Glyma.01G000600</th>\n",
       "      <th>...</th>\n",
       "      <th>Glyma.U044501</th>\n",
       "      <th>Glyma.U044700</th>\n",
       "      <th>Glyma.U044800</th>\n",
       "      <th>Glyma.U045000</th>\n",
       "      <th>Glyma.U045100</th>\n",
       "      <th>Glyma.U045402</th>\n",
       "      <th>Glyma.U045502</th>\n",
       "      <th>Glyma.U045602</th>\n",
       "      <th>Glyma.U045702</th>\n",
       "      <th>Glyma.U045802</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biosample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMN00265078</th>\n",
       "      <td>5.453108</td>\n",
       "      <td>59.054636</td>\n",
       "      <td>1.408092</td>\n",
       "      <td>1.218809</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.651369</td>\n",
       "      <td>2.801494</td>\n",
       "      <td>0.377480</td>\n",
       "      <td>0.523319</td>\n",
       "      <td>2.993846</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.317951</td>\n",
       "      <td>0.421438</td>\n",
       "      <td>11.072744</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.053070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN00265079</th>\n",
       "      <td>6.500151</td>\n",
       "      <td>56.896441</td>\n",
       "      <td>6.155613</td>\n",
       "      <td>2.903028</td>\n",
       "      <td>14.714201</td>\n",
       "      <td>12.752819</td>\n",
       "      <td>4.904175</td>\n",
       "      <td>6.120868</td>\n",
       "      <td>7.059885</td>\n",
       "      <td>6.651526</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.313916</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.977273</td>\n",
       "      <td>0.518904</td>\n",
       "      <td>1.336911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN00265080</th>\n",
       "      <td>3.103596</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.038168</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.434158</td>\n",
       "      <td>7.362351</td>\n",
       "      <td>1.693301</td>\n",
       "      <td>4.158352</td>\n",
       "      <td>9.958392</td>\n",
       "      <td>9.768484</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.723192</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.653363</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.470588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN00265081</th>\n",
       "      <td>4.226230</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.021155</td>\n",
       "      <td>0.441538</td>\n",
       "      <td>2.820968</td>\n",
       "      <td>4.016449</td>\n",
       "      <td>1.627095</td>\n",
       "      <td>4.222031</td>\n",
       "      <td>7.327622</td>\n",
       "      <td>10.667464</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.931953</td>\n",
       "      <td>0.297342</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.390688</td>\n",
       "      <td>0.777894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN00265082</th>\n",
       "      <td>2.582089</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.710374</td>\n",
       "      <td>0.980639</td>\n",
       "      <td>5.646594</td>\n",
       "      <td>3.689870</td>\n",
       "      <td>1.837848</td>\n",
       "      <td>8.406095</td>\n",
       "      <td>5.689184</td>\n",
       "      <td>4.331051</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.495017</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.276112</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.595369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN09295126</th>\n",
       "      <td>6.011819</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.234152</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.596353</td>\n",
       "      <td>1.998173</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.105045</td>\n",
       "      <td>4.742704</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.817960</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>32.194774</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN09295127</th>\n",
       "      <td>5.552961</td>\n",
       "      <td>21.092150</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.700031</td>\n",
       "      <td>1.792143</td>\n",
       "      <td>8.356985</td>\n",
       "      <td>3.354179</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.273037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.604034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN09295128</th>\n",
       "      <td>5.692426</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.120328</td>\n",
       "      <td>4.756147</td>\n",
       "      <td>3.019845</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.545678</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN09295129</th>\n",
       "      <td>3.511474</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.860411</td>\n",
       "      <td>1.673083</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.202417</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN09295130</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.040881</td>\n",
       "      <td>1.218035</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.632265</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5376 rows × 52837 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene          Glyma.01G000100  Glyma.01G000137  Glyma.01G000174  \\\n",
       "biosample                                                         \n",
       "SAMN00265078         5.453108        59.054636         1.408092   \n",
       "SAMN00265079         6.500151        56.896441         6.155613   \n",
       "SAMN00265080         3.103596         0.000000         4.038168   \n",
       "SAMN00265081         4.226230         0.000000         4.021155   \n",
       "SAMN00265082         2.582089         0.000000         1.710374   \n",
       "...                       ...              ...              ...   \n",
       "SAMN09295126         6.011819         0.000000         2.234152   \n",
       "SAMN09295127         5.552961        21.092150         0.000000   \n",
       "SAMN09295128         5.692426         0.000000         0.000000   \n",
       "SAMN09295129         3.511474         0.000000         0.000000   \n",
       "SAMN09295130         0.000000         0.000000         0.000000   \n",
       "\n",
       "Gene          Glyma.01G000211  Glyma.01G000248  Glyma.01G000285  \\\n",
       "biosample                                                         \n",
       "SAMN00265078         1.218809         0.000000         2.651369   \n",
       "SAMN00265079         2.903028        14.714201        12.752819   \n",
       "SAMN00265080         0.000000         2.434158         7.362351   \n",
       "SAMN00265081         0.441538         2.820968         4.016449   \n",
       "SAMN00265082         0.980639         5.646594         3.689870   \n",
       "...                       ...              ...              ...   \n",
       "SAMN09295126         0.000000         0.000000         2.596353   \n",
       "SAMN09295127         0.000000         0.000000         0.000000   \n",
       "SAMN09295128         0.000000         0.000000         0.000000   \n",
       "SAMN09295129         0.000000         0.000000         0.000000   \n",
       "SAMN09295130         0.000000         0.000000         0.000000   \n",
       "\n",
       "Gene          Glyma.01G000322  Glyma.01G000359  Glyma.01G000400  \\\n",
       "biosample                                                         \n",
       "SAMN00265078         2.801494         0.377480         0.523319   \n",
       "SAMN00265079         4.904175         6.120868         7.059885   \n",
       "SAMN00265080         1.693301         4.158352         9.958392   \n",
       "SAMN00265081         1.627095         4.222031         7.327622   \n",
       "SAMN00265082         1.837848         8.406095         5.689184   \n",
       "...                       ...              ...              ...   \n",
       "SAMN09295126         1.998173         0.000000         9.105045   \n",
       "SAMN09295127         0.700031         1.792143         8.356985   \n",
       "SAMN09295128         0.000000        13.120328         4.756147   \n",
       "SAMN09295129         0.000000         0.000000         1.860411   \n",
       "SAMN09295130         0.000000         0.000000         1.040881   \n",
       "\n",
       "Gene          Glyma.01G000600  ...  Glyma.U044501  Glyma.U044700  \\\n",
       "biosample                      ...                                 \n",
       "SAMN00265078         2.993846  ...            0.0       0.000000   \n",
       "SAMN00265079         6.651526  ...            0.0       0.000000   \n",
       "SAMN00265080         9.768484  ...            0.0       0.000000   \n",
       "SAMN00265081        10.667464  ...            0.0       0.000000   \n",
       "SAMN00265082         4.331051  ...            0.0       0.000000   \n",
       "...                       ...  ...            ...            ...   \n",
       "SAMN09295126         4.742704  ...            0.0       0.000000   \n",
       "SAMN09295127         3.354179  ...            0.0       0.000000   \n",
       "SAMN09295128         3.019845  ...            0.0       0.000000   \n",
       "SAMN09295129         1.673083  ...            0.0       2.202417   \n",
       "SAMN09295130         1.218035  ...            0.0       0.000000   \n",
       "\n",
       "Gene          Glyma.U044800  Glyma.U045000  Glyma.U045100  Glyma.U045402  \\\n",
       "biosample                                                                  \n",
       "SAMN00265078            0.0            0.0            0.0       4.317951   \n",
       "SAMN00265079            0.0            0.0            0.0       4.313916   \n",
       "SAMN00265080            0.0            0.0            0.0       0.723192   \n",
       "SAMN00265081            0.0            0.0            0.0       0.931953   \n",
       "SAMN00265082            0.0            0.0            0.0       0.495017   \n",
       "...                     ...            ...            ...            ...   \n",
       "SAMN09295126            0.0            0.0            0.0       8.817960   \n",
       "SAMN09295127            0.0            0.0            0.0       5.273037   \n",
       "SAMN09295128            0.0            0.0            0.0       1.545678   \n",
       "SAMN09295129            0.0            0.0            0.0       0.000000   \n",
       "SAMN09295130            0.0            0.0            0.0       0.000000   \n",
       "\n",
       "Gene          Glyma.U045502  Glyma.U045602  Glyma.U045702  Glyma.U045802  \n",
       "biosample                                                                 \n",
       "SAMN00265078       0.421438      11.072744       0.000000       0.053070  \n",
       "SAMN00265079       0.000000       8.977273       0.518904       1.336911  \n",
       "SAMN00265080       0.000000       6.653363       0.000000       2.470588  \n",
       "SAMN00265081       0.297342       0.000000       0.390688       0.777894  \n",
       "SAMN00265082       0.000000       2.276112       0.000000       0.595369  \n",
       "...                     ...            ...            ...            ...  \n",
       "SAMN09295126       0.000000      32.194774       0.000000       0.000000  \n",
       "SAMN09295127       0.000000       0.000000       0.000000       0.604034  \n",
       "SAMN09295128       0.000000       0.000000       0.000000       0.000000  \n",
       "SAMN09295129       0.000000       0.000000       0.000000       0.000000  \n",
       "SAMN09295130       0.000000       0.000000       0.000000       0.632265  \n",
       "\n",
       "[5376 rows x 52837 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "venancio = pd.read_parquet('../expresion/venancio.parquet')\n",
    "venancio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "condicion = pd.read_csv('../anotacion/condicion.csv', index_col=0)\n",
    "condicion = condicion[condicion.isna().sum(axis=1) == 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_194114/2108350401.py:1: DtypeWarning: Columns (60,69,72,78,95,96,97,103,104,105,109,112,116,121,122,132,133,138,139,141,145,146,147,161,166,167,169,170,171,173,174,176,177,179,180,186,187,188,189,190,191,192,193,194,195,196,197,199,200,201,202,203,204,205,207,212,213,218,219,220,222,224,225,226,227,228,229,230,231,232,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255,256,257,258,259,260,261,262,263,264,265,266) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  sra = pd.read_csv('../metadata/sra.csv', index_col=0)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>center_name</th>\n",
       "      <th>consent</th>\n",
       "      <th>experiment</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>instrument</th>\n",
       "      <th>librarylayout</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>...</th>\n",
       "      <th>quality_scoring_system_run</th>\n",
       "      <th>accession_ecotype_sam</th>\n",
       "      <th>tissue_and_developmental_stage_sam</th>\n",
       "      <th>experimental_factor__1_4_exp</th>\n",
       "      <th>concentration_sam</th>\n",
       "      <th>cultivation_pattren_sam</th>\n",
       "      <th>er_gfp_sam</th>\n",
       "      <th>subsrc_note_sam_s_dpl392</th>\n",
       "      <th>authors_sam</th>\n",
       "      <th>env_biome_sam</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biosample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMD00235537</th>\n",
       "      <td>DRR237204</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>RIKEN_CSRS</td>\n",
       "      <td>public</td>\n",
       "      <td>DRX227388</td>\n",
       "      <td>SAMD00235537</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PolyA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN30889879</th>\n",
       "      <td>SRR21660183</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>INNER MONGOLIA ACADEMY OF AGRICULTURAL &amp; ANIMA...</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX17659913</td>\n",
       "      <td>CK (1,2,3,4)</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN09513405</th>\n",
       "      <td>SRR7447776</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>THE OHIO STATE UNIVERSITY</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX4318240</td>\n",
       "      <td>SL209648</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PolyA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN09078292</th>\n",
       "      <td>SRR7124263</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX4045331</td>\n",
       "      <td>GSM3132253</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36760700</th>\n",
       "      <td>SRR25459671</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SOYBEAN</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX21192507</td>\n",
       "      <td>R89-2-1</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PCR</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN25827181</th>\n",
       "      <td>SRR17968854</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>HUAZHONG AGRICULTURAL UNIVERSITY</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX14125189</td>\n",
       "      <td>II-Early-12dpi-2C_rep2</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Oligo-dT</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN20971793</th>\n",
       "      <td>SRR15610895</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX11908208</td>\n",
       "      <td>GSM5536784</td>\n",
       "      <td>AB SOLiD System 3.0</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN41726398</th>\n",
       "      <td>SRR29315715</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NORTHEAST AGRICULTURAL UNIVERSITY</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX24832506</td>\n",
       "      <td>HPHO_MM_2</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>RT-PCR</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN07202264</th>\n",
       "      <td>SRR5655591</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX2892180</td>\n",
       "      <td>GSM2651534</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN09716352</th>\n",
       "      <td>SRR7592877</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX4457873</td>\n",
       "      <td>GSM3306085</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10992 rows × 266 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      acc assay_type  \\\n",
       "biosample                              \n",
       "SAMD00235537    DRR237204    RNA-Seq   \n",
       "SAMN30889879  SRR21660183    RNA-Seq   \n",
       "SAMN09513405   SRR7447776    RNA-Seq   \n",
       "SAMN09078292   SRR7124263    RNA-Seq   \n",
       "SAMN36760700  SRR25459671    RNA-Seq   \n",
       "...                   ...        ...   \n",
       "SAMN25827181  SRR17968854    RNA-Seq   \n",
       "SAMN20971793  SRR15610895    RNA-Seq   \n",
       "SAMN41726398  SRR29315715    RNA-Seq   \n",
       "SAMN07202264   SRR5655591    RNA-Seq   \n",
       "SAMN09716352   SRR7592877    RNA-Seq   \n",
       "\n",
       "                                                    center_name consent  \\\n",
       "biosample                                                                 \n",
       "SAMD00235537                                         RIKEN_CSRS  public   \n",
       "SAMN30889879  INNER MONGOLIA ACADEMY OF AGRICULTURAL & ANIMA...  public   \n",
       "SAMN09513405                          THE OHIO STATE UNIVERSITY  public   \n",
       "SAMN09078292                                                GEO  public   \n",
       "SAMN36760700                                            SOYBEAN  public   \n",
       "...                                                         ...     ...   \n",
       "SAMN25827181                   HUAZHONG AGRICULTURAL UNIVERSITY  public   \n",
       "SAMN20971793                                                GEO  public   \n",
       "SAMN41726398                  NORTHEAST AGRICULTURAL UNIVERSITY  public   \n",
       "SAMN07202264                                                GEO  public   \n",
       "SAMN09716352                                                GEO  public   \n",
       "\n",
       "               experiment             sample_name             instrument  \\\n",
       "biosample                                                                  \n",
       "SAMD00235537    DRX227388            SAMD00235537    Illumina HiSeq 2500   \n",
       "SAMN30889879  SRX17659913            CK (1,2,3,4)  Illumina NovaSeq 6000   \n",
       "SAMN09513405   SRX4318240                SL209648    Illumina HiSeq 2500   \n",
       "SAMN09078292   SRX4045331              GSM3132253    Illumina HiSeq 2000   \n",
       "SAMN36760700  SRX21192507                 R89-2-1    Illumina HiSeq 4000   \n",
       "...                   ...                     ...                    ...   \n",
       "SAMN25827181  SRX14125189  II-Early-12dpi-2C_rep2    Illumina HiSeq 4000   \n",
       "SAMN20971793  SRX11908208              GSM5536784    AB SOLiD System 3.0   \n",
       "SAMN41726398  SRX24832506               HPHO_MM_2    Illumina HiSeq 2000   \n",
       "SAMN07202264   SRX2892180              GSM2651534    Illumina HiSeq 2500   \n",
       "SAMN09716352   SRX4457873              GSM3306085    Illumina HiSeq 4000   \n",
       "\n",
       "             librarylayout libraryselection   librarysource  ...  \\\n",
       "biosample                                                    ...   \n",
       "SAMD00235537        PAIRED            PolyA  TRANSCRIPTOMIC  ...   \n",
       "SAMN30889879        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "SAMN09513405        PAIRED            PolyA  TRANSCRIPTOMIC  ...   \n",
       "SAMN09078292        SINGLE             cDNA  TRANSCRIPTOMIC  ...   \n",
       "SAMN36760700        PAIRED              PCR  TRANSCRIPTOMIC  ...   \n",
       "...                    ...              ...             ...  ...   \n",
       "SAMN25827181        PAIRED         Oligo-dT  TRANSCRIPTOMIC  ...   \n",
       "SAMN20971793        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "SAMN41726398        PAIRED           RT-PCR  TRANSCRIPTOMIC  ...   \n",
       "SAMN07202264        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "SAMN09716352        SINGLE             cDNA  TRANSCRIPTOMIC  ...   \n",
       "\n",
       "             quality_scoring_system_run accession_ecotype_sam  \\\n",
       "biosample                                                       \n",
       "SAMD00235537                        NaN                   NaN   \n",
       "SAMN30889879                        NaN                   NaN   \n",
       "SAMN09513405                        NaN                   NaN   \n",
       "SAMN09078292                        NaN                   NaN   \n",
       "SAMN36760700                        NaN                   NaN   \n",
       "...                                 ...                   ...   \n",
       "SAMN25827181                        NaN                   NaN   \n",
       "SAMN20971793                        NaN                   NaN   \n",
       "SAMN41726398                        NaN                   NaN   \n",
       "SAMN07202264                        NaN                   NaN   \n",
       "SAMN09716352                        NaN                   NaN   \n",
       "\n",
       "             tissue_and_developmental_stage_sam experimental_factor__1_4_exp  \\\n",
       "biosample                                                                      \n",
       "SAMD00235537                                NaN                          NaN   \n",
       "SAMN30889879                                NaN                          NaN   \n",
       "SAMN09513405                                NaN                          NaN   \n",
       "SAMN09078292                                NaN                          NaN   \n",
       "SAMN36760700                                NaN                          NaN   \n",
       "...                                         ...                          ...   \n",
       "SAMN25827181                                NaN                          NaN   \n",
       "SAMN20971793                                NaN                          NaN   \n",
       "SAMN41726398                                NaN                          NaN   \n",
       "SAMN07202264                                NaN                          NaN   \n",
       "SAMN09716352                                NaN                          NaN   \n",
       "\n",
       "             concentration_sam cultivation_pattren_sam  er_gfp_sam  \\\n",
       "biosample                                                            \n",
       "SAMD00235537               NaN                     NaN         NaN   \n",
       "SAMN30889879               NaN                     NaN         NaN   \n",
       "SAMN09513405               NaN                     NaN         NaN   \n",
       "SAMN09078292               NaN                     NaN         NaN   \n",
       "SAMN36760700               NaN                     NaN         NaN   \n",
       "...                        ...                     ...         ...   \n",
       "SAMN25827181               NaN                     NaN         NaN   \n",
       "SAMN20971793               NaN                     NaN         NaN   \n",
       "SAMN41726398               NaN                     NaN         NaN   \n",
       "SAMN07202264               NaN                     NaN         NaN   \n",
       "SAMN09716352               NaN                     NaN         NaN   \n",
       "\n",
       "              subsrc_note_sam_s_dpl392  authors_sam  env_biome_sam  \n",
       "biosample                                                           \n",
       "SAMD00235537                       NaN          NaN            NaN  \n",
       "SAMN30889879                       NaN          NaN            NaN  \n",
       "SAMN09513405                       NaN          NaN            NaN  \n",
       "SAMN09078292                       NaN          NaN            NaN  \n",
       "SAMN36760700                       NaN          NaN            NaN  \n",
       "...                                ...          ...            ...  \n",
       "SAMN25827181                       NaN          NaN            NaN  \n",
       "SAMN20971793                       NaN          NaN            NaN  \n",
       "SAMN41726398                       NaN          NaN            NaN  \n",
       "SAMN07202264                       NaN          NaN            NaN  \n",
       "SAMN09716352                       NaN          NaN            NaN  \n",
       "\n",
       "[10992 rows x 266 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sra = pd.read_csv('../metadata/sra.csv', index_col=0)\n",
    "sra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "condicion, venancio = condicion.align(venancio, join='inner', axis=0)\n",
    "sra, condicion = sra.align(condicion, join='right', axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preprocesamiento"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn import preprocessing, manifold\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_los_tres(X, condicion, sra):\n",
    "    fig, ax = plt.subplots(1, 3, figsize=(20,5))\n",
    "    for bioproject in sra['bioproject'].unique():\n",
    "        ax[0].scatter(X[sra['bioproject'] == bioproject, 0], X[sra['bioproject'] == bioproject, 1], alpha=0.5, label=condicion)\n",
    "        ax[0].set_title('bioproject')\n",
    "    for tejido in condicion['tejido'].unique():\n",
    "        ax[1].scatter(X[condicion['tejido'] == tejido, 0], X[condicion['tejido'] == tejido, 1], alpha=0.5, label=tejido)\n",
    "        ax[1].set_title('tejido')\n",
    "    for estres in condicion['estres'].unique():\n",
    "        ax[2].scatter(X[condicion['estres'] == estres, 0], X[condicion['estres'] == estres, 1], alpha=0.5, label=tejido)\n",
    "        ax[2].set_title('estres')\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "tsne = manifold.TSNE()\n",
    "venancio_tsne = tsne.fit_transform(venancio)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "plot_los_tres() missing 1 required positional argument: 'sra'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mplot_los_tres\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvenancio_tsne\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcondicion\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m plt\u001b[38;5;241m.\u001b[39msuptitle(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mt-SNE expresión con batch effect\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mTypeError\u001b[0m: plot_los_tres() missing 1 required positional argument: 'sra'"
     ]
    }
   ],
   "source": [
    "plot_los_tres(venancio_tsne, condicion, sra)\n",
    "plt.suptitle('t-SNE expresión con batch effect')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "venancio = venancio.groupby(sra['bioproject']).apply(lambda g: pd.DataFrame(preprocessing.scale(g), index=g.index, columns=g.columns)).droplevel('bioproject')\n",
    "venancio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "venancio_tsne = tsne.fit_transform(venancio)\n",
    "plot_los_tres(venancio_tsne, condicion, sra)\n",
    "plt.suptitle('t-SNE expresión \"sin\" batch effect')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Modelado"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn import model_selection, ensemble"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X = venancio\n",
    "y = condicion['estres']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y == 'control')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "random_forest = ensemble.RandomForestClassifier()\n",
    "params = {\n",
    "    'n_estimators': np.arange(100, 1000, 100),\n",
    "    'max_depth': np.arange(20, 50, 10),\n",
    "    'min_samples_split': np.arange(2, 15, 4),\n",
    "    'min_samples_leaf': np.arange(2, 15, 4),\n",
    "    'max_features': ['sqrt', 'log2'],\n",
    "    'bootstrap': [True]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "search = model_selection.RandomizedSearchCV(estimator=random_forest, param_distributions=params, n_jobs=-1)\n",
    "search.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "search.cv_results_['mean_test_score'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "estimator = search.best_estimator_\n",
    "model_selection.LearningCurveDisplay.from_estimator(estimator, X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "importances = np.flip(np.sort(estimator.feature_importances_))\n",
    "plt.plot(importances)\n",
    "plt.title('Importancias de features ordenadas (Random Forest)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(estimator.feature_importances_ > 0.002).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_tsne = tsne.fit_transform(X.loc[:, estimator.feature_importances_ > 0.002])\n",
    "plot_los_tres(X_tsne, condicion, sra)\n",
    "plt.suptitle('t-SNE expresión de genes seleccionados')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gbr = ensemble.GradientBoostingClassifier()\n",
    "params = {\n",
    "    \"loss\": ['log_loss', 'exponential'],\n",
    "    \"learning_rate\": np.arange(0, 10, 1),\n",
    "    \"n_estimators\": np.arange(1, 200, 1),\n",
    "    \"max_depth\": np.arange(1, 50, 1),\n",
    "    \"max_features\": ['sqrt', 'log2']\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "search = model_selection.RandomizedSearchCV(gbr, param_distributions=params, n_jobs=-1)\n",
    "search.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "search.cv_results_['mean_test_score'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_selection.LearningCurveDisplay.from_estimator(search.best_estimator_, X_train, y_train)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
