{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_33347/3173754459.py:1: DtypeWarning: Columns (28,29,30,49,51,54,55,56,57,59,60,61,62,63,65,66,67,71,72,73,74,75,76,78,79,80,81,82,83,84,85,87,88,89,91,92,93,94,96,98,99,100,102,103,104,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,135,136,137,138,139,140,141,142,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,162,163,164,166,167,168,169,170,171,172,173,174,175,176,177,179,180,182,183,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255,256,257,258,259,260,261,262,263,264,265,266,267,268,269) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  metadata = pd.read_csv('../data/joined.csv', index_col=0)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>center_name</th>\n",
       "      <th>consent</th>\n",
       "      <th>experiment</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>instrument</th>\n",
       "      <th>librarylayout</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>...</th>\n",
       "      <th>er_gfp_sam</th>\n",
       "      <th>subsrc_note_sam_s_dpl392</th>\n",
       "      <th>authors_sam</th>\n",
       "      <th>env_biome_sam</th>\n",
       "      <th>title</th>\n",
       "      <th>Identifiers</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Attributes</th>\n",
       "      <th>Accession</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biosample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMD00025071</th>\n",
       "      <td>DRR029571</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NUGSS</td>\n",
       "      <td>public</td>\n",
       "      <td>DRX026629</td>\n",
       "      <td>DRS020767</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Error: uid 25071 cannot get document summary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD00025072</th>\n",
       "      <td>DRR029572</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NUGSS</td>\n",
       "      <td>public</td>\n",
       "      <td>DRX026630</td>\n",
       "      <td>DRS020768</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Error: uid 25072 cannot get document summary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD00025073</th>\n",
       "      <td>DRR029573</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NUGSS</td>\n",
       "      <td>public</td>\n",
       "      <td>DRX026631</td>\n",
       "      <td>DRS020769</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Error: uid 25073 cannot get document summary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD00025074</th>\n",
       "      <td>DRR029574</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NUGSS</td>\n",
       "      <td>public</td>\n",
       "      <td>DRX026632</td>\n",
       "      <td>DRS020770</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Error: uid 25074 cannot get document summary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD00025075</th>\n",
       "      <td>DRR029575</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NUGSS</td>\n",
       "      <td>public</td>\n",
       "      <td>DRX026633</td>\n",
       "      <td>DRS020771</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Error: uid 25075 cannot get document summary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN41829221</th>\n",
       "      <td>SRR29409034</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>KANSAS STATE UNIVERSITY</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX24922515</td>\n",
       "      <td>Soybean308</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Plant sample from Glycine max</td>\n",
       "      <td>{'BioSample': 'SAMN41829221', 'Sample name': '...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'cultivar': 'Pharaoh', 'tissue': 'Stem'}</td>\n",
       "      <td>SAMN41829221</td>\n",
       "      <td>41829221.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN41829222</th>\n",
       "      <td>SRR29409033</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>KANSAS STATE UNIVERSITY</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX24922516</td>\n",
       "      <td>Soybean309</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Plant sample from Glycine max</td>\n",
       "      <td>{'BioSample': 'SAMN41829222', 'Sample name': '...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'cultivar': 'Pharaoh', 'tissue': 'Stem'}</td>\n",
       "      <td>SAMN41829222</td>\n",
       "      <td>41829222.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN41829223</th>\n",
       "      <td>SRR29409032</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>KANSAS STATE UNIVERSITY</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX24922517</td>\n",
       "      <td>Soybean310</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Plant sample from Glycine max</td>\n",
       "      <td>{'BioSample': 'SAMN41829223', 'Sample name': '...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'cultivar': 'Pharaoh', 'tissue': 'Stem'}</td>\n",
       "      <td>SAMN41829223</td>\n",
       "      <td>41829223.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN41829224</th>\n",
       "      <td>SRR29409031</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>KANSAS STATE UNIVERSITY</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX24922518</td>\n",
       "      <td>Soybean311</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Plant sample from Glycine max</td>\n",
       "      <td>{'BioSample': 'SAMN41829224', 'Sample name': '...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'cultivar': 'Dt974290', 'tissue': 'Stem'}</td>\n",
       "      <td>SAMN41829224</td>\n",
       "      <td>41829224.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN41829225</th>\n",
       "      <td>SRR29409030</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>KANSAS STATE UNIVERSITY</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX24922519</td>\n",
       "      <td>Soybean312</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Plant sample from Glycine max</td>\n",
       "      <td>{'BioSample': 'SAMN41829225', 'Sample name': '...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'cultivar': 'Dt974290', 'tissue': 'Stem'}</td>\n",
       "      <td>SAMN41829225</td>\n",
       "      <td>41829225.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10992 rows × 275 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      acc assay_type              center_name consent  \\\n",
       "biosample                                                               \n",
       "SAMD00025071    DRR029571    RNA-Seq                    NUGSS  public   \n",
       "SAMD00025072    DRR029572    RNA-Seq                    NUGSS  public   \n",
       "SAMD00025073    DRR029573    RNA-Seq                    NUGSS  public   \n",
       "SAMD00025074    DRR029574    RNA-Seq                    NUGSS  public   \n",
       "SAMD00025075    DRR029575    RNA-Seq                    NUGSS  public   \n",
       "...                   ...        ...                      ...     ...   \n",
       "SAMN41829221  SRR29409034    RNA-Seq  KANSAS STATE UNIVERSITY  public   \n",
       "SAMN41829222  SRR29409033    RNA-Seq  KANSAS STATE UNIVERSITY  public   \n",
       "SAMN41829223  SRR29409032    RNA-Seq  KANSAS STATE UNIVERSITY  public   \n",
       "SAMN41829224  SRR29409031    RNA-Seq  KANSAS STATE UNIVERSITY  public   \n",
       "SAMN41829225  SRR29409030    RNA-Seq  KANSAS STATE UNIVERSITY  public   \n",
       "\n",
       "               experiment sample_name                    instrument  \\\n",
       "biosample                                                             \n",
       "SAMD00025071    DRX026629   DRS020767  Illumina Genome Analyzer IIx   \n",
       "SAMD00025072    DRX026630   DRS020768  Illumina Genome Analyzer IIx   \n",
       "SAMD00025073    DRX026631   DRS020769  Illumina Genome Analyzer IIx   \n",
       "SAMD00025074    DRX026632   DRS020770  Illumina Genome Analyzer IIx   \n",
       "SAMD00025075    DRX026633   DRS020771  Illumina Genome Analyzer IIx   \n",
       "...                   ...         ...                           ...   \n",
       "SAMN41829221  SRX24922515  Soybean308                   NextSeq 500   \n",
       "SAMN41829222  SRX24922516  Soybean309                   NextSeq 500   \n",
       "SAMN41829223  SRX24922517  Soybean310                   NextSeq 500   \n",
       "SAMN41829224  SRX24922518  Soybean311                   NextSeq 500   \n",
       "SAMN41829225  SRX24922519  Soybean312                   NextSeq 500   \n",
       "\n",
       "             librarylayout libraryselection   librarysource  ... er_gfp_sam  \\\n",
       "biosample                                                    ...              \n",
       "SAMD00025071        SINGLE           RANDOM  TRANSCRIPTOMIC  ...        NaN   \n",
       "SAMD00025072        SINGLE           RANDOM  TRANSCRIPTOMIC  ...        NaN   \n",
       "SAMD00025073        SINGLE           RANDOM  TRANSCRIPTOMIC  ...        NaN   \n",
       "SAMD00025074        SINGLE           RANDOM  TRANSCRIPTOMIC  ...        NaN   \n",
       "SAMD00025075        SINGLE           RANDOM  TRANSCRIPTOMIC  ...        NaN   \n",
       "...                    ...              ...             ...  ...        ...   \n",
       "SAMN41829221        SINGLE             cDNA  TRANSCRIPTOMIC  ...        NaN   \n",
       "SAMN41829222        SINGLE             cDNA  TRANSCRIPTOMIC  ...        NaN   \n",
       "SAMN41829223        SINGLE             cDNA  TRANSCRIPTOMIC  ...        NaN   \n",
       "SAMN41829224        SINGLE             cDNA  TRANSCRIPTOMIC  ...        NaN   \n",
       "SAMN41829225        SINGLE             cDNA  TRANSCRIPTOMIC  ...        NaN   \n",
       "\n",
       "             subsrc_note_sam_s_dpl392 authors_sam env_biome_sam  \\\n",
       "biosample                                                         \n",
       "SAMD00025071                      NaN         NaN           NaN   \n",
       "SAMD00025072                      NaN         NaN           NaN   \n",
       "SAMD00025073                      NaN         NaN           NaN   \n",
       "SAMD00025074                      NaN         NaN           NaN   \n",
       "SAMD00025075                      NaN         NaN           NaN   \n",
       "...                               ...         ...           ...   \n",
       "SAMN41829221                      NaN         NaN           NaN   \n",
       "SAMN41829222                      NaN         NaN           NaN   \n",
       "SAMN41829223                      NaN         NaN           NaN   \n",
       "SAMN41829224                      NaN         NaN           NaN   \n",
       "SAMN41829225                      NaN         NaN           NaN   \n",
       "\n",
       "                                                     title  \\\n",
       "biosample                                                    \n",
       "SAMD00025071  Error: uid 25071 cannot get document summary   \n",
       "SAMD00025072  Error: uid 25072 cannot get document summary   \n",
       "SAMD00025073  Error: uid 25073 cannot get document summary   \n",
       "SAMD00025074  Error: uid 25074 cannot get document summary   \n",
       "SAMD00025075  Error: uid 25075 cannot get document summary   \n",
       "...                                                    ...   \n",
       "SAMN41829221                 Plant sample from Glycine max   \n",
       "SAMN41829222                 Plant sample from Glycine max   \n",
       "SAMN41829223                 Plant sample from Glycine max   \n",
       "SAMN41829224                 Plant sample from Glycine max   \n",
       "SAMN41829225                 Plant sample from Glycine max   \n",
       "\n",
       "                                                    Identifiers     Organism  \\\n",
       "biosample                                                                      \n",
       "SAMD00025071                                                NaN          NaN   \n",
       "SAMD00025072                                                NaN          NaN   \n",
       "SAMD00025073                                                NaN          NaN   \n",
       "SAMD00025074                                                NaN          NaN   \n",
       "SAMD00025075                                                NaN          NaN   \n",
       "...                                                         ...          ...   \n",
       "SAMN41829221  {'BioSample': 'SAMN41829221', 'Sample name': '...  Glycine max   \n",
       "SAMN41829222  {'BioSample': 'SAMN41829222', 'Sample name': '...  Glycine max   \n",
       "SAMN41829223  {'BioSample': 'SAMN41829223', 'Sample name': '...  Glycine max   \n",
       "SAMN41829224  {'BioSample': 'SAMN41829224', 'Sample name': '...  Glycine max   \n",
       "SAMN41829225  {'BioSample': 'SAMN41829225', 'Sample name': '...  Glycine max   \n",
       "\n",
       "                                              Attributes     Accession  \\\n",
       "biosample                                                                \n",
       "SAMD00025071                                         NaN           NaN   \n",
       "SAMD00025072                                         NaN           NaN   \n",
       "SAMD00025073                                         NaN           NaN   \n",
       "SAMD00025074                                         NaN           NaN   \n",
       "SAMD00025075                                         NaN           NaN   \n",
       "...                                                  ...           ...   \n",
       "SAMN41829221   {'cultivar': 'Pharaoh', 'tissue': 'Stem'}  SAMN41829221   \n",
       "SAMN41829222   {'cultivar': 'Pharaoh', 'tissue': 'Stem'}  SAMN41829222   \n",
       "SAMN41829223   {'cultivar': 'Pharaoh', 'tissue': 'Stem'}  SAMN41829223   \n",
       "SAMN41829224  {'cultivar': 'Dt974290', 'tissue': 'Stem'}  SAMN41829224   \n",
       "SAMN41829225  {'cultivar': 'Dt974290', 'tissue': 'Stem'}  SAMN41829225   \n",
       "\n",
       "                      ID  \n",
       "biosample                 \n",
       "SAMD00025071         NaN  \n",
       "SAMD00025072         NaN  \n",
       "SAMD00025073         NaN  \n",
       "SAMD00025074         NaN  \n",
       "SAMD00025075         NaN  \n",
       "...                  ...  \n",
       "SAMN41829221  41829221.0  \n",
       "SAMN41829222  41829222.0  \n",
       "SAMN41829223  41829223.0  \n",
       "SAMN41829224  41829224.0  \n",
       "SAMN41829225  41829225.0  \n",
       "\n",
       "[10992 rows x 275 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_csv('../data/joined.csv', index_col=0)\n",
    "metadata\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def matching_cols(row, pat):\n",
    "    return row[row.astype(str).str.contains(pat, regex=False, case=False).fillna(False)].to_dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Junto palabras clave en kws. Guardo ```{columna: valor}``` en ```estreses``` si la columna contiene la palabra clave en su valor."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "kws = ['control', 'ctrl', 'stress', 'drought', 'bacteria', 'inoculated', 'flood', 'deficit', 'infest', 'pathogen', 'acid', 'fung', 'sporus']\n",
    "estreses = pd.DataFrame(index = metadata.index)\n",
    "\n",
    "for kw in kws:\n",
    "    estreses[kw] = metadata.apply(lambda r: matching_cols(r, kw), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>control</th>\n",
       "      <th>ctrl</th>\n",
       "      <th>stress</th>\n",
       "      <th>drought</th>\n",
       "      <th>bacteria</th>\n",
       "      <th>inoculated</th>\n",
       "      <th>flood</th>\n",
       "      <th>deficit</th>\n",
       "      <th>infest</th>\n",
       "      <th>pathogen</th>\n",
       "      <th>acid</th>\n",
       "      <th>fung</th>\n",
       "      <th>sporus</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biosample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMD00025071</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD00025072</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD00025073</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD00025074</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD00025075</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN41829221</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN41829222</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN41829223</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN41829224</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN41829225</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10992 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              control   ctrl  stress  drought  bacteria  inoculated  flood  \\\n",
       "biosample                                                                    \n",
       "SAMD00025071    False  False   False    False     False       False  False   \n",
       "SAMD00025072    False  False   False    False     False       False  False   \n",
       "SAMD00025073    False  False   False    False     False       False  False   \n",
       "SAMD00025074    False  False   False    False     False       False  False   \n",
       "SAMD00025075    False  False   False    False     False       False  False   \n",
       "...               ...    ...     ...      ...       ...         ...    ...   \n",
       "SAMN41829221    False  False   False    False     False       False  False   \n",
       "SAMN41829222    False  False   False    False     False       False  False   \n",
       "SAMN41829223    False  False   False    False     False       False  False   \n",
       "SAMN41829224    False  False   False    False     False       False  False   \n",
       "SAMN41829225    False  False   False    False     False       False  False   \n",
       "\n",
       "              deficit  infest  pathogen   acid   fung  sporus  \n",
       "biosample                                                      \n",
       "SAMD00025071    False   False     False  False  False   False  \n",
       "SAMD00025072    False   False     False  False  False   False  \n",
       "SAMD00025073    False   False     False  False  False   False  \n",
       "SAMD00025074    False   False     False  False  False   False  \n",
       "SAMD00025075    False   False     False  False  False   False  \n",
       "...               ...     ...       ...    ...    ...     ...  \n",
       "SAMN41829221    False   False     False  False  False   False  \n",
       "SAMN41829222    False   False     False  False  False   False  \n",
       "SAMN41829223    False   False     False  False  False   False  \n",
       "SAMN41829224    False   False     False  False  False   False  \n",
       "SAMN41829225    False   False     False  False  False   False  \n",
       "\n",
       "[10992 rows x 13 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indicadores = estreses.map(lambda v: v != {})\n",
    "indicadores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "control       441\n",
       "ctrl           13\n",
       "stress        485\n",
       "drought        84\n",
       "bacteria        0\n",
       "inoculated    339\n",
       "flood           0\n",
       "deficit       135\n",
       "infest        141\n",
       "pathogen        3\n",
       "acid           79\n",
       "fung           12\n",
       "sporus          9\n",
       "dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indicadores.sum(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "biosample\n",
       "SAMD00025071    0\n",
       "SAMD00025072    0\n",
       "SAMD00025073    0\n",
       "SAMD00025074    0\n",
       "SAMD00025075    0\n",
       "               ..\n",
       "SAMN41829221    0\n",
       "SAMN41829222    0\n",
       "SAMN41829223    0\n",
       "SAMN41829224    0\n",
       "SAMN41829225    0\n",
       "Length: 10992, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indicadores.sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "estreses = estreses[indicadores.sum(axis=1) > 0]\n",
    "indicadores = indicadores.loc[estreses.index]\n",
    "estreses.to_csv('estreses.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Condición\n",
    "## Estres vs. control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cols_true(row):\n",
    "# Devuelve lista de columnas verdaderas para row\n",
    "    return row[row].index.to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "condicion = pd.DataFrame(index = estreses.index)\n",
    "condicion.loc[indicadores['control'] | indicadores['ctrl'], 'condicion'] = 'control'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stress</th>\n",
       "      <th>n</th>\n",
       "      <th>part</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bioproject</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PRJNA250761</th>\n",
       "      <td>control</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown: 1</td>\n",
       "      <td>Glycine max strain:williams82 Transcriptome or...</td>\n",
       "      <td>JGI Flagship Plant Gene Atlas</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRJNA251047</th>\n",
       "      <td>control</td>\n",
       "      <td>1</td>\n",
       "      <td>unknown: 1</td>\n",
       "      <td>Glycine max strain:Williams 82 Transcriptome o...</td>\n",
       "      <td>JGI Flagship Plant Gene Atlas</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRJNA271453</th>\n",
       "      <td>control</td>\n",
       "      <td>3</td>\n",
       "      <td>leaf: 2 | root: 1</td>\n",
       "      <td>Glycine max Transcriptome or Gene expression</td>\n",
       "      <td>Transcriptome sequencing of Glycine max C08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRJNA271506</th>\n",
       "      <td>control</td>\n",
       "      <td>2</td>\n",
       "      <td>flower: 2</td>\n",
       "      <td>Glycine max Transcriptome or Gene expression</td>\n",
       "      <td>soybean Transcriptome or Gene expression</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRJNA339653</th>\n",
       "      <td>control</td>\n",
       "      <td>4</td>\n",
       "      <td>leaf: 4</td>\n",
       "      <td>Glycine max Raw sequence reads</td>\n",
       "      <td>RNA-seq of soybean accessions</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRJNA668549</th>\n",
       "      <td>control</td>\n",
       "      <td>8</td>\n",
       "      <td>leaf: 8</td>\n",
       "      <td>Soybean Transcriptome (clean reads)</td>\n",
       "      <td>Discover SMV-resistance genes/networks in soyb...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRJNA758180</th>\n",
       "      <td>control</td>\n",
       "      <td>1</td>\n",
       "      <td>hypocotyl: 1</td>\n",
       "      <td>Glycine max Raw sequence reads</td>\n",
       "      <td>Glycine max RNA-seq reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRJNA801060</th>\n",
       "      <td>control</td>\n",
       "      <td>18</td>\n",
       "      <td>seed: 18</td>\n",
       "      <td>Glycine max Raw sequence reads</td>\n",
       "      <td>normal RNA-Seq of soybean seeds</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRJNA389558</th>\n",
       "      <td>control</td>\n",
       "      <td>102</td>\n",
       "      <td>leaf: 102</td>\n",
       "      <td>Transcriptomes of 102 soybean accessions</td>\n",
       "      <td>A comparison of RNA expression profiles across...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRJNA692674</th>\n",
       "      <td>control</td>\n",
       "      <td>6</td>\n",
       "      <td>leaf: 6</td>\n",
       "      <td>RNAseq analysis of pooled soybean leaf samples...</td>\n",
       "      <td>Leaf samples were collected from soybean produ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>154 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              stress    n               part  \\\n",
       "bioproject                                     \n",
       "PRJNA250761  control    1         unknown: 1   \n",
       "PRJNA251047  control    1         unknown: 1   \n",
       "PRJNA271453  control    3  leaf: 2 | root: 1   \n",
       "PRJNA271506  control    2          flower: 2   \n",
       "PRJNA339653  control    4            leaf: 4   \n",
       "...              ...  ...                ...   \n",
       "PRJNA668549  control    8            leaf: 8   \n",
       "PRJNA758180  control    1       hypocotyl: 1   \n",
       "PRJNA801060  control   18           seed: 18   \n",
       "PRJNA389558  control  102          leaf: 102   \n",
       "PRJNA692674  control    6            leaf: 6   \n",
       "\n",
       "                                                         title  \\\n",
       "bioproject                                                       \n",
       "PRJNA250761  Glycine max strain:williams82 Transcriptome or...   \n",
       "PRJNA251047  Glycine max strain:Williams 82 Transcriptome o...   \n",
       "PRJNA271453       Glycine max Transcriptome or Gene expression   \n",
       "PRJNA271506       Glycine max Transcriptome or Gene expression   \n",
       "PRJNA339653                     Glycine max Raw sequence reads   \n",
       "...                                                        ...   \n",
       "PRJNA668549                Soybean Transcriptome (clean reads)   \n",
       "PRJNA758180                     Glycine max Raw sequence reads   \n",
       "PRJNA801060                     Glycine max Raw sequence reads   \n",
       "PRJNA389558           Transcriptomes of 102 soybean accessions   \n",
       "PRJNA692674  RNAseq analysis of pooled soybean leaf samples...   \n",
       "\n",
       "                                                      abstract  doi pmid  \n",
       "bioproject                                                                \n",
       "PRJNA250761                      JGI Flagship Plant Gene Atlas  NaN  NaN  \n",
       "PRJNA251047                      JGI Flagship Plant Gene Atlas  NaN  NaN  \n",
       "PRJNA271453        Transcriptome sequencing of Glycine max C08  NaN  NaN  \n",
       "PRJNA271506           soybean Transcriptome or Gene expression  NaN  NaN  \n",
       "PRJNA339653                      RNA-seq of soybean accessions  NaN  NaN  \n",
       "...                                                        ...  ...  ...  \n",
       "PRJNA668549  Discover SMV-resistance genes/networks in soyb...  NaN  NaN  \n",
       "PRJNA758180                          Glycine max RNA-seq reads  NaN  NaN  \n",
       "PRJNA801060                    normal RNA-Seq of soybean seeds  NaN  NaN  \n",
       "PRJNA389558  A comparison of RNA expression profiles across...  NaN  NaN  \n",
       "PRJNA692674  Leaf samples were collected from soybean produ...  NaN  NaN  \n",
       "\n",
       "[154 rows x 7 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bioprojects = pd.read_csv('../metadata/venancio.csv', index_col=0)\n",
    "bioprojects[bioprojects['stress'] == 'control']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_stress_project = metadata.reset_index().set_index('bioproject').join(bioprojects, rsuffix='_bioprj').set_index('biosample')['stress']\n",
    "condicion.loc[sample_stress_project == 'control', 'condicion'] = 'control'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "control = condicion['condicion'] == 'control'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Estrés"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condicion</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biosample</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMD00133783</th>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD00133784</th>\n",
       "      <td>[drought]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD00133785</th>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD00133786</th>\n",
       "      <td>[drought]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD00133787</th>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN40876398</th>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN40876399</th>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN40876400</th>\n",
       "      <td>[stress]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN40876401</th>\n",
       "      <td>[stress]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN40876402</th>\n",
       "      <td>[stress]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1574 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              condicion\n",
       "biosample              \n",
       "SAMD00133783    control\n",
       "SAMD00133784  [drought]\n",
       "SAMD00133785    control\n",
       "SAMD00133786  [drought]\n",
       "SAMD00133787    control\n",
       "...                 ...\n",
       "SAMN40876398    control\n",
       "SAMN40876399    control\n",
       "SAMN40876400   [stress]\n",
       "SAMN40876401   [stress]\n",
       "SAMN40876402   [stress]\n",
       "\n",
       "[1574 rows x 1 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "condicion.loc[~control, 'condicion'] = indicadores[~control].apply(cols_true, axis = 1)\n",
    "condicion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condicion</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biosample</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMD00222779</th>\n",
       "      <td>[inoculated, pathogen]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN03755595</th>\n",
       "      <td>[stress, deficit]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN03755596</th>\n",
       "      <td>[stress, deficit]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN03755597</th>\n",
       "      <td>[stress, deficit]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN03755601</th>\n",
       "      <td>[stress, deficit]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN38032658</th>\n",
       "      <td>[stress, drought]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN38032659</th>\n",
       "      <td>[stress, drought]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN38032660</th>\n",
       "      <td>[stress, drought]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN38032661</th>\n",
       "      <td>[stress, drought]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN38032662</th>\n",
       "      <td>[stress, drought]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           condicion\n",
       "biosample                           \n",
       "SAMD00222779  [inoculated, pathogen]\n",
       "SAMN03755595       [stress, deficit]\n",
       "SAMN03755596       [stress, deficit]\n",
       "SAMN03755597       [stress, deficit]\n",
       "SAMN03755601       [stress, deficit]\n",
       "...                              ...\n",
       "SAMN38032658       [stress, drought]\n",
       "SAMN38032659       [stress, drought]\n",
       "SAMN38032660       [stress, drought]\n",
       "SAMN38032661       [stress, drought]\n",
       "SAMN38032662       [stress, drought]\n",
       "\n",
       "[100 rows x 1 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multi = ~control & (condicion.loc[~control, 'condicion'].map(len) > 1)\n",
    "condicion[multi]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Muchas muestras tienen stress además del estrés específico. Me quedo con este último."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condicion</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biosample</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMD00222779</th>\n",
       "      <td>[inoculated, pathogen]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN03755595</th>\n",
       "      <td>[deficit]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN03755596</th>\n",
       "      <td>[deficit]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN03755597</th>\n",
       "      <td>[deficit]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN03755601</th>\n",
       "      <td>[deficit]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN38032658</th>\n",
       "      <td>[drought]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN38032659</th>\n",
       "      <td>[drought]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN38032660</th>\n",
       "      <td>[drought]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN38032661</th>\n",
       "      <td>[drought]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN38032662</th>\n",
       "      <td>[drought]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           condicion\n",
       "biosample                           \n",
       "SAMD00222779  [inoculated, pathogen]\n",
       "SAMN03755595               [deficit]\n",
       "SAMN03755596               [deficit]\n",
       "SAMN03755597               [deficit]\n",
       "SAMN03755601               [deficit]\n",
       "...                              ...\n",
       "SAMN38032658               [drought]\n",
       "SAMN38032659               [drought]\n",
       "SAMN38032660               [drought]\n",
       "SAMN38032661               [drought]\n",
       "SAMN38032662               [drought]\n",
       "\n",
       "[100 rows x 1 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "condicion.loc[multi, 'condicion'] = condicion.loc[multi, 'condicion'].apply(lambda r: [s for s in r if s != 'stress'])\n",
    "condicion[multi]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Repito a ver cuales quedaron con varios estreses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condicion</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biosample</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMD00222779</th>\n",
       "      <td>[inoculated, pathogen]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN08469975</th>\n",
       "      <td>[inoculated, pathogen]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN08469977</th>\n",
       "      <td>[inoculated, pathogen]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN09295122</th>\n",
       "      <td>[inoculated, fung]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN09295123</th>\n",
       "      <td>[inoculated, fung]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN09295124</th>\n",
       "      <td>[inoculated, fung]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN09295128</th>\n",
       "      <td>[inoculated, fung]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN09295129</th>\n",
       "      <td>[inoculated, fung]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN09295130</th>\n",
       "      <td>[inoculated, fung]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630190</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630191</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630192</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630193</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630194</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630195</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630196</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630197</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630198</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630205</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630206</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630207</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630214</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630215</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630216</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630244</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630245</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630246</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630247</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630248</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630249</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630250</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630251</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630252</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630259</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630260</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630261</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630268</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630269</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36630270</th>\n",
       "      <td>[deficit, acid]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           condicion\n",
       "biosample                           \n",
       "SAMD00222779  [inoculated, pathogen]\n",
       "SAMN08469975  [inoculated, pathogen]\n",
       "SAMN08469977  [inoculated, pathogen]\n",
       "SAMN09295122      [inoculated, fung]\n",
       "SAMN09295123      [inoculated, fung]\n",
       "SAMN09295124      [inoculated, fung]\n",
       "SAMN09295128      [inoculated, fung]\n",
       "SAMN09295129      [inoculated, fung]\n",
       "SAMN09295130      [inoculated, fung]\n",
       "SAMN36630190         [deficit, acid]\n",
       "SAMN36630191         [deficit, acid]\n",
       "SAMN36630192         [deficit, acid]\n",
       "SAMN36630193         [deficit, acid]\n",
       "SAMN36630194         [deficit, acid]\n",
       "SAMN36630195         [deficit, acid]\n",
       "SAMN36630196         [deficit, acid]\n",
       "SAMN36630197         [deficit, acid]\n",
       "SAMN36630198         [deficit, acid]\n",
       "SAMN36630205         [deficit, acid]\n",
       "SAMN36630206         [deficit, acid]\n",
       "SAMN36630207         [deficit, acid]\n",
       "SAMN36630214         [deficit, acid]\n",
       "SAMN36630215         [deficit, acid]\n",
       "SAMN36630216         [deficit, acid]\n",
       "SAMN36630244         [deficit, acid]\n",
       "SAMN36630245         [deficit, acid]\n",
       "SAMN36630246         [deficit, acid]\n",
       "SAMN36630247         [deficit, acid]\n",
       "SAMN36630248         [deficit, acid]\n",
       "SAMN36630249         [deficit, acid]\n",
       "SAMN36630250         [deficit, acid]\n",
       "SAMN36630251         [deficit, acid]\n",
       "SAMN36630252         [deficit, acid]\n",
       "SAMN36630259         [deficit, acid]\n",
       "SAMN36630260         [deficit, acid]\n",
       "SAMN36630261         [deficit, acid]\n",
       "SAMN36630268         [deficit, acid]\n",
       "SAMN36630269         [deficit, acid]\n",
       "SAMN36630270         [deficit, acid]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multi = ~control & (condicion.loc[~control, 'condicion'].map(len) > 1)\n",
    "condicion[multi]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "<lambda>() got an unexpected keyword argument 'axis'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[20], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m condicion\u001b[38;5;241m.\u001b[39mloc[multi, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcondicion\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43mcondicion\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloc\u001b[49m\u001b[43m[\u001b[49m\u001b[43mmulti\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mcondicion\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43;01mlambda\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mr\u001b[49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[43ms\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43ms\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mr\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43ms\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m!=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mdeficit\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mand\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43ms\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m!=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43minoculated\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m condicion[multi]\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/series.py:4904\u001b[0m, in \u001b[0;36mSeries.apply\u001b[0;34m(self, func, convert_dtype, args, by_row, **kwargs)\u001b[0m\n\u001b[1;32m   4769\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mapply\u001b[39m(\n\u001b[1;32m   4770\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   4771\u001b[0m     func: AggFuncType,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   4776\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m   4777\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DataFrame \u001b[38;5;241m|\u001b[39m Series:\n\u001b[1;32m   4778\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   4779\u001b[0m \u001b[38;5;124;03m    Invoke function on values of Series.\u001b[39;00m\n\u001b[1;32m   4780\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   4895\u001b[0m \u001b[38;5;124;03m    dtype: float64\u001b[39;00m\n\u001b[1;32m   4896\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m   4897\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mSeriesApply\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   4898\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4899\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4900\u001b[0m \u001b[43m        \u001b[49m\u001b[43mconvert_dtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconvert_dtype\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4901\u001b[0m \u001b[43m        \u001b[49m\u001b[43mby_row\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mby_row\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4902\u001b[0m \u001b[43m        \u001b[49m\u001b[43margs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4903\u001b[0m \u001b[43m        \u001b[49m\u001b[43mkwargs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m-> 4904\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/apply.py:1427\u001b[0m, in \u001b[0;36mSeriesApply.apply\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1424\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mapply_compat()\n\u001b[1;32m   1426\u001b[0m \u001b[38;5;66;03m# self.func is Callable\u001b[39;00m\n\u001b[0;32m-> 1427\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply_standard\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/apply.py:1507\u001b[0m, in \u001b[0;36mSeriesApply.apply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1501\u001b[0m \u001b[38;5;66;03m# row-wise access\u001b[39;00m\n\u001b[1;32m   1502\u001b[0m \u001b[38;5;66;03m# apply doesn't have a `na_action` keyword and for backward compat reasons\u001b[39;00m\n\u001b[1;32m   1503\u001b[0m \u001b[38;5;66;03m# we need to give `na_action=\"ignore\"` for categorical data.\u001b[39;00m\n\u001b[1;32m   1504\u001b[0m \u001b[38;5;66;03m# TODO: remove the `na_action=\"ignore\"` when that default has been changed in\u001b[39;00m\n\u001b[1;32m   1505\u001b[0m \u001b[38;5;66;03m#  Categorical (GH51645).\u001b[39;00m\n\u001b[1;32m   1506\u001b[0m action \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mignore\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(obj\u001b[38;5;241m.\u001b[39mdtype, CategoricalDtype) \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1507\u001b[0m mapped \u001b[38;5;241m=\u001b[39m \u001b[43mobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_map_values\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1508\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmapper\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcurried\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mna_action\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maction\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconvert\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconvert_dtype\u001b[49m\n\u001b[1;32m   1509\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1511\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(mapped) \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(mapped[\u001b[38;5;241m0\u001b[39m], ABCSeries):\n\u001b[1;32m   1512\u001b[0m     \u001b[38;5;66;03m# GH#43986 Need to do list(mapped) in order to get treated as nested\u001b[39;00m\n\u001b[1;32m   1513\u001b[0m     \u001b[38;5;66;03m#  See also GH#25959 regarding EA support\u001b[39;00m\n\u001b[1;32m   1514\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m obj\u001b[38;5;241m.\u001b[39m_constructor_expanddim(\u001b[38;5;28mlist\u001b[39m(mapped), index\u001b[38;5;241m=\u001b[39mobj\u001b[38;5;241m.\u001b[39mindex)\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/base.py:921\u001b[0m, in \u001b[0;36mIndexOpsMixin._map_values\u001b[0;34m(self, mapper, na_action, convert)\u001b[0m\n\u001b[1;32m    918\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(arr, ExtensionArray):\n\u001b[1;32m    919\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m arr\u001b[38;5;241m.\u001b[39mmap(mapper, na_action\u001b[38;5;241m=\u001b[39mna_action)\n\u001b[0;32m--> 921\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43malgorithms\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmap_array\u001b[49m\u001b[43m(\u001b[49m\u001b[43marr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmapper\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mna_action\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mna_action\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconvert\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconvert\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/algorithms.py:1743\u001b[0m, in \u001b[0;36mmap_array\u001b[0;34m(arr, mapper, na_action, convert)\u001b[0m\n\u001b[1;32m   1741\u001b[0m values \u001b[38;5;241m=\u001b[39m arr\u001b[38;5;241m.\u001b[39mastype(\u001b[38;5;28mobject\u001b[39m, copy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m   1742\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m na_action \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 1743\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mlib\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmap_infer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmapper\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconvert\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconvert\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1744\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1745\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m lib\u001b[38;5;241m.\u001b[39mmap_infer_mask(\n\u001b[1;32m   1746\u001b[0m         values, mapper, mask\u001b[38;5;241m=\u001b[39misna(values)\u001b[38;5;241m.\u001b[39mview(np\u001b[38;5;241m.\u001b[39muint8), convert\u001b[38;5;241m=\u001b[39mconvert\n\u001b[1;32m   1747\u001b[0m     )\n",
      "File \u001b[0;32mlib.pyx:2972\u001b[0m, in \u001b[0;36mpandas._libs.lib.map_infer\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/apply.py:1496\u001b[0m, in \u001b[0;36mSeriesApply.apply_standard.<locals>.curried\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m   1495\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcurried\u001b[39m(x):\n\u001b[0;32m-> 1496\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mTypeError\u001b[0m: <lambda>() got an unexpected keyword argument 'axis'"
     ]
    }
   ],
   "source": [
    "condicion.loc[multi, 'condicion'] = condicion.loc[multi, 'condicion'].apply(lambda r: [s for s in r if s != 'deficit' and s != 'inoculated'])\n",
    "condicion[multi]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ahora los transformo a ```str```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isinstance(condicion.iloc[1,0], list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "condicion.to_csv('condicion.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
