{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re, json, requests\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "acc = pd.read_csv('../data/glycine_max_sra.csv')\n",
    "acc = acc.loc[acc['assay_type'] == 'RNA-Seq']\n",
    "acc = acc.set_index('biosample')\n",
    "acc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "acc.loc['SAMN30889879']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "attr = pd.DataFrame.from_records(acc['jattr'].apply(eval), index = acc.index)\n",
    "attr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "attr.isna().mean().sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "acc.to_csv('helper/acc.csv')\n",
    "attr.to_csv('helper/attr.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_biosample(biosamp_id):\n",
    "    api_url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=biosample&id={biosamp_id}&rettype=full&retmode=text\"\n",
    "    res = requests.get(api_url)\n",
    "    return res.text\n",
    "\n",
    "def metadata_json(metadata):\n",
    "    res = {}\n",
    "    res['title'] = metadata.split('\\n')[0].replace('1: ', '')\n",
    "    identifiers = re.search(r'Identifiers: (.+)', metadata)\n",
    "    if identifiers:\n",
    "        res['Identifiers'] = {k.strip(): v.strip() for k, v in \n",
    "                                [item.split(':') if ':' in item else (item, \"\")\n",
    "                                    for item in identifiers.group(1).split(';')]}\n",
    "    organism = re.search(r'Organism: (.+)', metadata)\n",
    "    if organism:\n",
    "        res['Organism'] = organism.group(1).strip()\n",
    "    attributes = re.findall(r'/(\\w+)=\"([^\"]+)\"', metadata)\n",
    "    if attributes:\n",
    "        res['Attributes'] = {k: v for k, v in attributes}\n",
    "    accession = re.search(r'Accession: (\\w+)', metadata)\n",
    "    if accession:\n",
    "        res['Accession'] = accession.group(1)\n",
    "    id_match = re.search(r'ID: (\\d+)', metadata)\n",
    "    if id_match:\n",
    "        res['ID'] = id_match.group(1)\n",
    "    \n",
    "    return json.dumps(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    biosamp = pd.read_csv('helper/biosamp.csv', index_col=0)\n",
    "except:\n",
    "    biosamp = pd.DataFrame(acc['biosample'])\n",
    "    biosamp['metadata'] = acc['biosample'].apply(get_biosample)\n",
    "    biosamp['jmeta'] = biosamp['metadata'].apply(metadata_json)\n",
    "    biosamp = biosamp.set_index('biosample')\n",
    "    biosamp = pd.DataFrame(biosamp)\n",
    "    biosamp.to_csv('helper/biosamp.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_18239/1722256263.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  jmeta = pd.DataFrame.from_records(biosamp['metadata'].apply(metadata_json).apply(json.loads), index = biosamp.index)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>Identifiers</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Attributes</th>\n",
       "      <th>Accession</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biosample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMD00235537</th>\n",
       "      <td>Oryza sativa Japonica Group NG8334</td>\n",
       "      <td>{'BioSample': 'SAMN00235537', 'GSS': 'LIBGSS_0...</td>\n",
       "      <td>Oryza sativa Japonica Group</td>\n",
       "      <td>{'strain': 'NG8334', 'cultivar': 'Nipponbare'}</td>\n",
       "      <td>SAMN00235537</td>\n",
       "      <td>235537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN30889879</th>\n",
       "      <td>Plant sample from Glycine max</td>\n",
       "      <td>{'BioSample': 'SAMN30889879', 'Sample name': '...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'isolate': 'B sufficient', 'cultivar': 'missi...</td>\n",
       "      <td>SAMN30889879</td>\n",
       "      <td>30889879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN09513405</th>\n",
       "      <td>Plant sample from Glycine max</td>\n",
       "      <td>{'BioSample': 'SAMN09513405', 'Sample name': '...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'isolate': 'P. sojae 1.S.1.1', 'cultivar': '1...</td>\n",
       "      <td>SAMN09513405</td>\n",
       "      <td>9513405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN09078292</th>\n",
       "      <td>mRNA gmax PI641156 NC-Raleigh</td>\n",
       "      <td>{'BioSample': 'SAMN09078292', 'SRA': 'SRS32627...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'cultivar': 'PI641156 NC-Raleigh', 'tissue': ...</td>\n",
       "      <td>SAMN09078292</td>\n",
       "      <td>9078292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN36760700</th>\n",
       "      <td>Plant sample from Glycine max</td>\n",
       "      <td>{'BioSample': 'SAMN36760700', 'Sample name': '...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'isolate': 'R89-2 line repeat 1 in CSSL popul...</td>\n",
       "      <td>SAMN36760700</td>\n",
       "      <td>36760700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN25827181</th>\n",
       "      <td>12dpi-2C-Replicate2</td>\n",
       "      <td>{'BioSample': 'SAMN25827181', 'Sample name': '...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'cultivar': 'Williams 82', 'age': '12 days po...</td>\n",
       "      <td>SAMN25827181</td>\n",
       "      <td>25827181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN20971793</th>\n",
       "      <td>Mock-inoculated Williams rep 2 [WM-2a]</td>\n",
       "      <td>{'BioSample': 'SAMN20971793', 'SRA': 'SRS99190...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'cultivar': 'Soybean cultivar Williams'}</td>\n",
       "      <td>SAMN20971793</td>\n",
       "      <td>20971793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN41726398</th>\n",
       "      <td>Plant sample from Glycine max</td>\n",
       "      <td>{'BioSample': 'SAMN41726398', 'Sample name': '...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'isolate': 'HPHO_2', 'tissue': 'seed'}</td>\n",
       "      <td>SAMN41726398</td>\n",
       "      <td>41726398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN07202264</th>\n",
       "      <td>LG92-1255_RNA-seq</td>\n",
       "      <td>{'BioSample': 'SAMN07202264', 'SRA': 'SRS22599...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'tissue': 'leaves', 'cultivar': 'LG92-1255'}</td>\n",
       "      <td>SAMN07202264</td>\n",
       "      <td>7202264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN09716352</th>\n",
       "      <td>YH03_3D4</td>\n",
       "      <td>{'BioSample': 'SAMN09716352', 'SRA': 'SRS35954...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'tissue': 'The whole shoot above cotelydon', ...</td>\n",
       "      <td>SAMN09716352</td>\n",
       "      <td>9716352</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10984 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               title  \\\n",
       "biosample                                              \n",
       "SAMD00235537      Oryza sativa Japonica Group NG8334   \n",
       "SAMN30889879           Plant sample from Glycine max   \n",
       "SAMN09513405           Plant sample from Glycine max   \n",
       "SAMN09078292           mRNA gmax PI641156 NC-Raleigh   \n",
       "SAMN36760700           Plant sample from Glycine max   \n",
       "...                                              ...   \n",
       "SAMN25827181                     12dpi-2C-Replicate2   \n",
       "SAMN20971793  Mock-inoculated Williams rep 2 [WM-2a]   \n",
       "SAMN41726398           Plant sample from Glycine max   \n",
       "SAMN07202264                       LG92-1255_RNA-seq   \n",
       "SAMN09716352                                YH03_3D4   \n",
       "\n",
       "                                                    Identifiers  \\\n",
       "biosample                                                         \n",
       "SAMD00235537  {'BioSample': 'SAMN00235537', 'GSS': 'LIBGSS_0...   \n",
       "SAMN30889879  {'BioSample': 'SAMN30889879', 'Sample name': '...   \n",
       "SAMN09513405  {'BioSample': 'SAMN09513405', 'Sample name': '...   \n",
       "SAMN09078292  {'BioSample': 'SAMN09078292', 'SRA': 'SRS32627...   \n",
       "SAMN36760700  {'BioSample': 'SAMN36760700', 'Sample name': '...   \n",
       "...                                                         ...   \n",
       "SAMN25827181  {'BioSample': 'SAMN25827181', 'Sample name': '...   \n",
       "SAMN20971793  {'BioSample': 'SAMN20971793', 'SRA': 'SRS99190...   \n",
       "SAMN41726398  {'BioSample': 'SAMN41726398', 'Sample name': '...   \n",
       "SAMN07202264  {'BioSample': 'SAMN07202264', 'SRA': 'SRS22599...   \n",
       "SAMN09716352  {'BioSample': 'SAMN09716352', 'SRA': 'SRS35954...   \n",
       "\n",
       "                                 Organism  \\\n",
       "biosample                                   \n",
       "SAMD00235537  Oryza sativa Japonica Group   \n",
       "SAMN30889879                  Glycine max   \n",
       "SAMN09513405                  Glycine max   \n",
       "SAMN09078292                  Glycine max   \n",
       "SAMN36760700                  Glycine max   \n",
       "...                                   ...   \n",
       "SAMN25827181                  Glycine max   \n",
       "SAMN20971793                  Glycine max   \n",
       "SAMN41726398                  Glycine max   \n",
       "SAMN07202264                  Glycine max   \n",
       "SAMN09716352                  Glycine max   \n",
       "\n",
       "                                                     Attributes     Accession  \\\n",
       "biosample                                                                       \n",
       "SAMD00235537     {'strain': 'NG8334', 'cultivar': 'Nipponbare'}  SAMN00235537   \n",
       "SAMN30889879  {'isolate': 'B sufficient', 'cultivar': 'missi...  SAMN30889879   \n",
       "SAMN09513405  {'isolate': 'P. sojae 1.S.1.1', 'cultivar': '1...  SAMN09513405   \n",
       "SAMN09078292  {'cultivar': 'PI641156 NC-Raleigh', 'tissue': ...  SAMN09078292   \n",
       "SAMN36760700  {'isolate': 'R89-2 line repeat 1 in CSSL popul...  SAMN36760700   \n",
       "...                                                         ...           ...   \n",
       "SAMN25827181  {'cultivar': 'Williams 82', 'age': '12 days po...  SAMN25827181   \n",
       "SAMN20971793          {'cultivar': 'Soybean cultivar Williams'}  SAMN20971793   \n",
       "SAMN41726398            {'isolate': 'HPHO_2', 'tissue': 'seed'}  SAMN41726398   \n",
       "SAMN07202264      {'tissue': 'leaves', 'cultivar': 'LG92-1255'}  SAMN07202264   \n",
       "SAMN09716352  {'tissue': 'The whole shoot above cotelydon', ...  SAMN09716352   \n",
       "\n",
       "                    ID  \n",
       "biosample               \n",
       "SAMD00235537    235537  \n",
       "SAMN30889879  30889879  \n",
       "SAMN09513405   9513405  \n",
       "SAMN09078292   9078292  \n",
       "SAMN36760700  36760700  \n",
       "...                ...  \n",
       "SAMN25827181  25827181  \n",
       "SAMN20971793  20971793  \n",
       "SAMN41726398  41726398  \n",
       "SAMN07202264   7202264  \n",
       "SAMN09716352   9716352  \n",
       "\n",
       "[10984 rows x 6 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "jmeta = pd.DataFrame.from_records(biosamp['metadata'].apply(metadata_json).apply(json.loads), index = biosamp.index)\n",
    "jmeta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_18239/987814684.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  jattr = pd.DataFrame.from_records(jmeta['Attributes'].dropna(), index = jmeta['Attributes'].dropna().index)\n",
      "/tmp/ipykernel_18239/987814684.py:2: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  jident = pd.DataFrame.from_records(jmeta['Identifiers'].dropna(), index = jmeta['Identifiers'].dropna().index).drop('BioSample', axis=1)\n"
     ]
    }
   ],
   "source": [
    "jattr = pd.DataFrame.from_records(jmeta['Attributes'].dropna(), index = jmeta['Attributes'].dropna().index)\n",
    "jident = pd.DataFrame.from_records(jmeta['Identifiers'].dropna(), index = jmeta['Identifiers'].dropna().index).drop('BioSample', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "biometa = pd.DataFrame(jmeta['title']).join(jattr, how='left', rsuffix='_jattr').join(jident, how='left', rsuffix='_jident')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "biometa.to_csv('helper/biometa.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Unified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_18239/2955333836.py:2: DtypeWarning: Columns (26,61,62,69,78,87,88,98,99,104,105,111,113,127,132,133,139,142,143,145,146,152,153,154,155,156,157,158,159,160,161,162,163,165,166,167,169,170,171,173,178,179,180,186,188,190,191,192,193,194,195,196,201,202,203,204,205,206,207,208,211,212,213,214,215,216,217,219,220,221,222,223,224,225,226,227,228,229,230,231,232) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  attr = pd.read_csv('helper/attr.csv')\n",
      "/tmp/ipykernel_18239/2955333836.py:3: DtypeWarning: Columns (27,31,63,64,66,85,87,88,89,92,93) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  biometa = pd.read_csv('helper/biometa.csv')\n"
     ]
    }
   ],
   "source": [
    "acc = pd.read_csv('helper/acc.csv')\n",
    "attr = pd.read_csv('helper/attr.csv')\n",
    "biometa = pd.read_csv('helper/biometa.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "acc = acc.drop_duplicates(keep='first', subset='biosample').set_index('biosample')\n",
    "attr = attr.drop_duplicates(keep='first', subset='biosample').set_index('biosample')\n",
    "biometa = biometa.drop_duplicates(keep='first', subset='biosample').set_index('biosample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "joined = acc.join(attr, how='outer', rsuffix='_attr').join(jmeta, how='outer', rsuffix='_jmeta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>center_name</th>\n",
       "      <th>consent</th>\n",
       "      <th>experiment</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>instrument</th>\n",
       "      <th>librarylayout</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>...</th>\n",
       "      <th>er_gfp_sam</th>\n",
       "      <th>subsrc_note_sam_s_dpl392</th>\n",
       "      <th>authors_sam</th>\n",
       "      <th>env_biome_sam</th>\n",
       "      <th>title</th>\n",
       "      <th>Identifiers</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Attributes</th>\n",
       "      <th>Accession</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biosample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMD00025071</th>\n",
       "      <td>DRR029571</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NUGSS</td>\n",
       "      <td>public</td>\n",
       "      <td>DRX026629</td>\n",
       "      <td>DRS020767</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Error: uid 25071 cannot get document summary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD00025072</th>\n",
       "      <td>DRR029572</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NUGSS</td>\n",
       "      <td>public</td>\n",
       "      <td>DRX026630</td>\n",
       "      <td>DRS020768</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Error: uid 25072 cannot get document summary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD00025073</th>\n",
       "      <td>DRR029573</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NUGSS</td>\n",
       "      <td>public</td>\n",
       "      <td>DRX026631</td>\n",
       "      <td>DRS020769</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Error: uid 25073 cannot get document summary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD00025074</th>\n",
       "      <td>DRR029574</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NUGSS</td>\n",
       "      <td>public</td>\n",
       "      <td>DRX026632</td>\n",
       "      <td>DRS020770</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Error: uid 25074 cannot get document summary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD00025075</th>\n",
       "      <td>DRR029575</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NUGSS</td>\n",
       "      <td>public</td>\n",
       "      <td>DRX026633</td>\n",
       "      <td>DRS020771</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Error: uid 25075 cannot get document summary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN41829221</th>\n",
       "      <td>SRR29409034</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>KANSAS STATE UNIVERSITY</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX24922515</td>\n",
       "      <td>Soybean308</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Plant sample from Glycine max</td>\n",
       "      <td>{'BioSample': 'SAMN41829221', 'Sample name': '...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'cultivar': 'Pharaoh', 'tissue': 'Stem'}</td>\n",
       "      <td>SAMN41829221</td>\n",
       "      <td>41829221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN41829222</th>\n",
       "      <td>SRR29409033</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>KANSAS STATE UNIVERSITY</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX24922516</td>\n",
       "      <td>Soybean309</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Plant sample from Glycine max</td>\n",
       "      <td>{'BioSample': 'SAMN41829222', 'Sample name': '...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'cultivar': 'Pharaoh', 'tissue': 'Stem'}</td>\n",
       "      <td>SAMN41829222</td>\n",
       "      <td>41829222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN41829223</th>\n",
       "      <td>SRR29409032</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>KANSAS STATE UNIVERSITY</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX24922517</td>\n",
       "      <td>Soybean310</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Plant sample from Glycine max</td>\n",
       "      <td>{'BioSample': 'SAMN41829223', 'Sample name': '...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'cultivar': 'Pharaoh', 'tissue': 'Stem'}</td>\n",
       "      <td>SAMN41829223</td>\n",
       "      <td>41829223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN41829224</th>\n",
       "      <td>SRR29409031</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>KANSAS STATE UNIVERSITY</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX24922518</td>\n",
       "      <td>Soybean311</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Plant sample from Glycine max</td>\n",
       "      <td>{'BioSample': 'SAMN41829224', 'Sample name': '...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'cultivar': 'Dt974290', 'tissue': 'Stem'}</td>\n",
       "      <td>SAMN41829224</td>\n",
       "      <td>41829224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMN41829225</th>\n",
       "      <td>SRR29409030</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>KANSAS STATE UNIVERSITY</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX24922519</td>\n",
       "      <td>Soybean312</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Plant sample from Glycine max</td>\n",
       "      <td>{'BioSample': 'SAMN41829225', 'Sample name': '...</td>\n",
       "      <td>Glycine max</td>\n",
       "      <td>{'cultivar': 'Dt974290', 'tissue': 'Stem'}</td>\n",
       "      <td>SAMN41829225</td>\n",
       "      <td>41829225</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10992 rows Ã— 275 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      acc assay_type              center_name consent  \\\n",
       "biosample                                                               \n",
       "SAMD00025071    DRR029571    RNA-Seq                    NUGSS  public   \n",
       "SAMD00025072    DRR029572    RNA-Seq                    NUGSS  public   \n",
       "SAMD00025073    DRR029573    RNA-Seq                    NUGSS  public   \n",
       "SAMD00025074    DRR029574    RNA-Seq                    NUGSS  public   \n",
       "SAMD00025075    DRR029575    RNA-Seq                    NUGSS  public   \n",
       "...                   ...        ...                      ...     ...   \n",
       "SAMN41829221  SRR29409034    RNA-Seq  KANSAS STATE UNIVERSITY  public   \n",
       "SAMN41829222  SRR29409033    RNA-Seq  KANSAS STATE UNIVERSITY  public   \n",
       "SAMN41829223  SRR29409032    RNA-Seq  KANSAS STATE UNIVERSITY  public   \n",
       "SAMN41829224  SRR29409031    RNA-Seq  KANSAS STATE UNIVERSITY  public   \n",
       "SAMN41829225  SRR29409030    RNA-Seq  KANSAS STATE UNIVERSITY  public   \n",
       "\n",
       "               experiment sample_name                    instrument  \\\n",
       "biosample                                                             \n",
       "SAMD00025071    DRX026629   DRS020767  Illumina Genome Analyzer IIx   \n",
       "SAMD00025072    DRX026630   DRS020768  Illumina Genome Analyzer IIx   \n",
       "SAMD00025073    DRX026631   DRS020769  Illumina Genome Analyzer IIx   \n",
       "SAMD00025074    DRX026632   DRS020770  Illumina Genome Analyzer IIx   \n",
       "SAMD00025075    DRX026633   DRS020771  Illumina Genome Analyzer IIx   \n",
       "...                   ...         ...                           ...   \n",
       "SAMN41829221  SRX24922515  Soybean308                   NextSeq 500   \n",
       "SAMN41829222  SRX24922516  Soybean309                   NextSeq 500   \n",
       "SAMN41829223  SRX24922517  Soybean310                   NextSeq 500   \n",
       "SAMN41829224  SRX24922518  Soybean311                   NextSeq 500   \n",
       "SAMN41829225  SRX24922519  Soybean312                   NextSeq 500   \n",
       "\n",
       "             librarylayout libraryselection   librarysource  ... er_gfp_sam  \\\n",
       "biosample                                                    ...              \n",
       "SAMD00025071        SINGLE           RANDOM  TRANSCRIPTOMIC  ...        NaN   \n",
       "SAMD00025072        SINGLE           RANDOM  TRANSCRIPTOMIC  ...        NaN   \n",
       "SAMD00025073        SINGLE           RANDOM  TRANSCRIPTOMIC  ...        NaN   \n",
       "SAMD00025074        SINGLE           RANDOM  TRANSCRIPTOMIC  ...        NaN   \n",
       "SAMD00025075        SINGLE           RANDOM  TRANSCRIPTOMIC  ...        NaN   \n",
       "...                    ...              ...             ...  ...        ...   \n",
       "SAMN41829221        SINGLE             cDNA  TRANSCRIPTOMIC  ...        NaN   \n",
       "SAMN41829222        SINGLE             cDNA  TRANSCRIPTOMIC  ...        NaN   \n",
       "SAMN41829223        SINGLE             cDNA  TRANSCRIPTOMIC  ...        NaN   \n",
       "SAMN41829224        SINGLE             cDNA  TRANSCRIPTOMIC  ...        NaN   \n",
       "SAMN41829225        SINGLE             cDNA  TRANSCRIPTOMIC  ...        NaN   \n",
       "\n",
       "             subsrc_note_sam_s_dpl392 authors_sam env_biome_sam  \\\n",
       "biosample                                                         \n",
       "SAMD00025071                      NaN         NaN           NaN   \n",
       "SAMD00025072                      NaN         NaN           NaN   \n",
       "SAMD00025073                      NaN         NaN           NaN   \n",
       "SAMD00025074                      NaN         NaN           NaN   \n",
       "SAMD00025075                      NaN         NaN           NaN   \n",
       "...                               ...         ...           ...   \n",
       "SAMN41829221                      NaN         NaN           NaN   \n",
       "SAMN41829222                      NaN         NaN           NaN   \n",
       "SAMN41829223                      NaN         NaN           NaN   \n",
       "SAMN41829224                      NaN         NaN           NaN   \n",
       "SAMN41829225                      NaN         NaN           NaN   \n",
       "\n",
       "                                                     title  \\\n",
       "biosample                                                    \n",
       "SAMD00025071  Error: uid 25071 cannot get document summary   \n",
       "SAMD00025072  Error: uid 25072 cannot get document summary   \n",
       "SAMD00025073  Error: uid 25073 cannot get document summary   \n",
       "SAMD00025074  Error: uid 25074 cannot get document summary   \n",
       "SAMD00025075  Error: uid 25075 cannot get document summary   \n",
       "...                                                    ...   \n",
       "SAMN41829221                 Plant sample from Glycine max   \n",
       "SAMN41829222                 Plant sample from Glycine max   \n",
       "SAMN41829223                 Plant sample from Glycine max   \n",
       "SAMN41829224                 Plant sample from Glycine max   \n",
       "SAMN41829225                 Plant sample from Glycine max   \n",
       "\n",
       "                                                    Identifiers     Organism  \\\n",
       "biosample                                                                      \n",
       "SAMD00025071                                                NaN          NaN   \n",
       "SAMD00025072                                                NaN          NaN   \n",
       "SAMD00025073                                                NaN          NaN   \n",
       "SAMD00025074                                                NaN          NaN   \n",
       "SAMD00025075                                                NaN          NaN   \n",
       "...                                                         ...          ...   \n",
       "SAMN41829221  {'BioSample': 'SAMN41829221', 'Sample name': '...  Glycine max   \n",
       "SAMN41829222  {'BioSample': 'SAMN41829222', 'Sample name': '...  Glycine max   \n",
       "SAMN41829223  {'BioSample': 'SAMN41829223', 'Sample name': '...  Glycine max   \n",
       "SAMN41829224  {'BioSample': 'SAMN41829224', 'Sample name': '...  Glycine max   \n",
       "SAMN41829225  {'BioSample': 'SAMN41829225', 'Sample name': '...  Glycine max   \n",
       "\n",
       "                                              Attributes     Accession  \\\n",
       "biosample                                                                \n",
       "SAMD00025071                                         NaN           NaN   \n",
       "SAMD00025072                                         NaN           NaN   \n",
       "SAMD00025073                                         NaN           NaN   \n",
       "SAMD00025074                                         NaN           NaN   \n",
       "SAMD00025075                                         NaN           NaN   \n",
       "...                                                  ...           ...   \n",
       "SAMN41829221   {'cultivar': 'Pharaoh', 'tissue': 'Stem'}  SAMN41829221   \n",
       "SAMN41829222   {'cultivar': 'Pharaoh', 'tissue': 'Stem'}  SAMN41829222   \n",
       "SAMN41829223   {'cultivar': 'Pharaoh', 'tissue': 'Stem'}  SAMN41829223   \n",
       "SAMN41829224  {'cultivar': 'Dt974290', 'tissue': 'Stem'}  SAMN41829224   \n",
       "SAMN41829225  {'cultivar': 'Dt974290', 'tissue': 'Stem'}  SAMN41829225   \n",
       "\n",
       "                    ID  \n",
       "biosample               \n",
       "SAMD00025071       NaN  \n",
       "SAMD00025072       NaN  \n",
       "SAMD00025073       NaN  \n",
       "SAMD00025074       NaN  \n",
       "SAMD00025075       NaN  \n",
       "...                ...  \n",
       "SAMN41829221  41829221  \n",
       "SAMN41829222  41829222  \n",
       "SAMN41829223  41829223  \n",
       "SAMN41829224  41829224  \n",
       "SAMN41829225  41829225  \n",
       "\n",
       "[10992 rows x 275 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# joined = pd.concat([acc, attr, jmeta], axis=1)\n",
    "joined.to_csv('joined.csv')\n",
    "joined"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bioprojects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bioproject(bioproj_id):\n",
    "    api_url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=bioproject&id={bioproj_id}\"\n",
    "    res = requests.get(api_url)\n",
    "    return res.text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bioproject</th>\n",
       "      <th>n</th>\n",
       "      <th>part</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA360609</td>\n",
       "      <td>8</td>\n",
       "      <td>seedling: 8</td>\n",
       "      <td>Transgenic and conventional cultivar compariso...</td>\n",
       "      <td>Soybean is one of the main sources of oil worl...</td>\n",
       "      <td>10.1105/tpc.18.00662</td>\n",
       "      <td>31227558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA369113</td>\n",
       "      <td>36</td>\n",
       "      <td>leaf: 36</td>\n",
       "      <td>Circadian transcriptome of soybean unifoliolat...</td>\n",
       "      <td>The goal of this study is to obtain the circad...</td>\n",
       "      <td>10.1073/pnas.1708508116</td>\n",
       "      <td>31676549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJNA369414</td>\n",
       "      <td>21</td>\n",
       "      <td>embryo: 15 | cotyledon: 6</td>\n",
       "      <td>Genome-Wide Reinforcement of DNA Methylation O...</td>\n",
       "      <td>Tissue culture is a major method for plant reg...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJNA369483</td>\n",
       "      <td>6</td>\n",
       "      <td>seed: 6</td>\n",
       "      <td>Characterizing seed weight related genes throu...</td>\n",
       "      <td>Cultivated soybean has domesticated in China f...</td>\n",
       "      <td>http://dx.doi.org/10.1007/s00299-017-2165-5</td>\n",
       "      <td>28653111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJNA372408</td>\n",
       "      <td>1</td>\n",
       "      <td>shoot: 1</td>\n",
       "      <td>Gene expression analysis of Glycine max willia...</td>\n",
       "      <td>Transcriptome sequencing as part of the JGI Fl...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>489</th>\n",
       "      <td>PRJNA827520</td>\n",
       "      <td>6</td>\n",
       "      <td>leaf: 6</td>\n",
       "      <td>Glycine max Transcriptome or Gene expression</td>\n",
       "      <td>Glycine max treate with water or GZM to test t...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>490</th>\n",
       "      <td>PRJNA827666</td>\n",
       "      <td>6</td>\n",
       "      <td>root: 6</td>\n",
       "      <td>Study on improvement of water and nutrient upt...</td>\n",
       "      <td>Application and/or natural Silicates (Si) is c...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>491</th>\n",
       "      <td>PRJNA832118</td>\n",
       "      <td>24</td>\n",
       "      <td>leaf: 24</td>\n",
       "      <td>Glycine max Transcriptome or Gene expression</td>\n",
       "      <td>Transgenic GmFT3a soybean</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492</th>\n",
       "      <td>PRJNA833532</td>\n",
       "      <td>12</td>\n",
       "      <td>root: 12</td>\n",
       "      <td>Glycine max Raw sequence reads</td>\n",
       "      <td>soybean aluminum stress</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493</th>\n",
       "      <td>PRJNA834849</td>\n",
       "      <td>6</td>\n",
       "      <td>leaf: 6</td>\n",
       "      <td>Soybean drought transcriptome</td>\n",
       "      <td>15% PEG-6000 simulated drought treatment</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>494 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      bioproject   n                       part  \\\n",
       "0    PRJNA360609   8                seedling: 8   \n",
       "1    PRJNA369113  36                   leaf: 36   \n",
       "2    PRJNA369414  21  embryo: 15 | cotyledon: 6   \n",
       "3    PRJNA369483   6                    seed: 6   \n",
       "4    PRJNA372408   1                   shoot: 1   \n",
       "..           ...  ..                        ...   \n",
       "489  PRJNA827520   6                    leaf: 6   \n",
       "490  PRJNA827666   6                    root: 6   \n",
       "491  PRJNA832118  24                   leaf: 24   \n",
       "492  PRJNA833532  12                   root: 12   \n",
       "493  PRJNA834849   6                    leaf: 6   \n",
       "\n",
       "                                                 title  \\\n",
       "0    Transgenic and conventional cultivar compariso...   \n",
       "1    Circadian transcriptome of soybean unifoliolat...   \n",
       "2    Genome-Wide Reinforcement of DNA Methylation O...   \n",
       "3    Characterizing seed weight related genes throu...   \n",
       "4    Gene expression analysis of Glycine max willia...   \n",
       "..                                                 ...   \n",
       "489       Glycine max Transcriptome or Gene expression   \n",
       "490  Study on improvement of water and nutrient upt...   \n",
       "491       Glycine max Transcriptome or Gene expression   \n",
       "492                     Glycine max Raw sequence reads   \n",
       "493                      Soybean drought transcriptome   \n",
       "\n",
       "                                              abstract  \\\n",
       "0    Soybean is one of the main sources of oil worl...   \n",
       "1    The goal of this study is to obtain the circad...   \n",
       "2    Tissue culture is a major method for plant reg...   \n",
       "3    Cultivated soybean has domesticated in China f...   \n",
       "4    Transcriptome sequencing as part of the JGI Fl...   \n",
       "..                                                 ...   \n",
       "489  Glycine max treate with water or GZM to test t...   \n",
       "490  Application and/or natural Silicates (Si) is c...   \n",
       "491                          Transgenic GmFT3a soybean   \n",
       "492                            soybean aluminum stress   \n",
       "493           15% PEG-6000 simulated drought treatment   \n",
       "\n",
       "                                             doi      pmid  \n",
       "0                           10.1105/tpc.18.00662  31227558  \n",
       "1                        10.1073/pnas.1708508116  31676549  \n",
       "2                                           None      None  \n",
       "3    http://dx.doi.org/10.1007/s00299-017-2165-5  28653111  \n",
       "4                                           None      None  \n",
       "..                                           ...       ...  \n",
       "489                                         None      None  \n",
       "490                                         None      None  \n",
       "491                                         None      None  \n",
       "492                                         None      None  \n",
       "493                                         None      None  \n",
       "\n",
       "[494 rows x 7 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "venancio = pd.read_json('venancio.json')\n",
    "venancio.columns = ['bioproject', 'n', 'part', 'title', 'abstract', 'doi', 'pmid']\n",
    "venancio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bioproject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJDB3474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJDB3582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJDB7011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJDB7219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJDB7775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>785</th>\n",
       "      <td>PRJNA1119938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>786</th>\n",
       "      <td>PRJNA1120606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>787</th>\n",
       "      <td>PRJNA1121049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>788</th>\n",
       "      <td>PRJNA1121077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789</th>\n",
       "      <td>PRJNA1123857</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>790 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       bioproject\n",
       "0       PRJDB3474\n",
       "1       PRJDB3582\n",
       "2       PRJDB7011\n",
       "3       PRJDB7219\n",
       "4       PRJDB7775\n",
       "..            ...\n",
       "785  PRJNA1119938\n",
       "786  PRJNA1120606\n",
       "787  PRJNA1121049\n",
       "788  PRJNA1121077\n",
       "789  PRJNA1123857\n",
       "\n",
       "[790 rows x 1 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "projects = pd.DataFrame({'bioproject': joined.bioproject.unique()})\n",
    "projects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bioproject</th>\n",
       "      <th>n</th>\n",
       "      <th>part</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJDB3474</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJDB3582</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJDB7011</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJDB7219</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJDB7775</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>785</th>\n",
       "      <td>PRJNA1119938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>786</th>\n",
       "      <td>PRJNA1120606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>787</th>\n",
       "      <td>PRJNA1121049</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>788</th>\n",
       "      <td>PRJNA1121077</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789</th>\n",
       "      <td>PRJNA1123857</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>790 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       bioproject   n part title abstract  doi pmid\n",
       "0       PRJDB3474 NaN  NaN   NaN      NaN  NaN  NaN\n",
       "1       PRJDB3582 NaN  NaN   NaN      NaN  NaN  NaN\n",
       "2       PRJDB7011 NaN  NaN   NaN      NaN  NaN  NaN\n",
       "3       PRJDB7219 NaN  NaN   NaN      NaN  NaN  NaN\n",
       "4       PRJDB7775 NaN  NaN   NaN      NaN  NaN  NaN\n",
       "..            ...  ..  ...   ...      ...  ...  ...\n",
       "785  PRJNA1119938 NaN  NaN   NaN      NaN  NaN  NaN\n",
       "786  PRJNA1120606 NaN  NaN   NaN      NaN  NaN  NaN\n",
       "787  PRJNA1121049 NaN  NaN   NaN      NaN  NaN  NaN\n",
       "788  PRJNA1121077 NaN  NaN   NaN      NaN  NaN  NaN\n",
       "789  PRJNA1123857 NaN  NaN   NaN      NaN  NaN  NaN\n",
       "\n",
       "[790 rows x 7 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "projects = projects.join(venancio.set_index('bioproject'), rsuffix='_venancio')\n",
    "projects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[23], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m projects[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mxml\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43mprojects\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mbioproject\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mget_bioproject\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/series.py:4904\u001b[0m, in \u001b[0;36mSeries.apply\u001b[0;34m(self, func, convert_dtype, args, by_row, **kwargs)\u001b[0m\n\u001b[1;32m   4769\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mapply\u001b[39m(\n\u001b[1;32m   4770\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   4771\u001b[0m     func: AggFuncType,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   4776\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m   4777\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DataFrame \u001b[38;5;241m|\u001b[39m Series:\n\u001b[1;32m   4778\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   4779\u001b[0m \u001b[38;5;124;03m    Invoke function on values of Series.\u001b[39;00m\n\u001b[1;32m   4780\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   4895\u001b[0m \u001b[38;5;124;03m    dtype: float64\u001b[39;00m\n\u001b[1;32m   4896\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m   4897\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mSeriesApply\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   4898\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4899\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4900\u001b[0m \u001b[43m        \u001b[49m\u001b[43mconvert_dtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconvert_dtype\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4901\u001b[0m \u001b[43m        \u001b[49m\u001b[43mby_row\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mby_row\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4902\u001b[0m \u001b[43m        \u001b[49m\u001b[43margs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   4903\u001b[0m \u001b[43m        \u001b[49m\u001b[43mkwargs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m-> 4904\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/apply.py:1427\u001b[0m, in \u001b[0;36mSeriesApply.apply\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1424\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mapply_compat()\n\u001b[1;32m   1426\u001b[0m \u001b[38;5;66;03m# self.func is Callable\u001b[39;00m\n\u001b[0;32m-> 1427\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply_standard\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/apply.py:1507\u001b[0m, in \u001b[0;36mSeriesApply.apply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1501\u001b[0m \u001b[38;5;66;03m# row-wise access\u001b[39;00m\n\u001b[1;32m   1502\u001b[0m \u001b[38;5;66;03m# apply doesn't have a `na_action` keyword and for backward compat reasons\u001b[39;00m\n\u001b[1;32m   1503\u001b[0m \u001b[38;5;66;03m# we need to give `na_action=\"ignore\"` for categorical data.\u001b[39;00m\n\u001b[1;32m   1504\u001b[0m \u001b[38;5;66;03m# TODO: remove the `na_action=\"ignore\"` when that default has been changed in\u001b[39;00m\n\u001b[1;32m   1505\u001b[0m \u001b[38;5;66;03m#  Categorical (GH51645).\u001b[39;00m\n\u001b[1;32m   1506\u001b[0m action \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mignore\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(obj\u001b[38;5;241m.\u001b[39mdtype, CategoricalDtype) \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1507\u001b[0m mapped \u001b[38;5;241m=\u001b[39m \u001b[43mobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_map_values\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1508\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmapper\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcurried\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mna_action\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maction\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconvert\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconvert_dtype\u001b[49m\n\u001b[1;32m   1509\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1511\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(mapped) \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(mapped[\u001b[38;5;241m0\u001b[39m], ABCSeries):\n\u001b[1;32m   1512\u001b[0m     \u001b[38;5;66;03m# GH#43986 Need to do list(mapped) in order to get treated as nested\u001b[39;00m\n\u001b[1;32m   1513\u001b[0m     \u001b[38;5;66;03m#  See also GH#25959 regarding EA support\u001b[39;00m\n\u001b[1;32m   1514\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m obj\u001b[38;5;241m.\u001b[39m_constructor_expanddim(\u001b[38;5;28mlist\u001b[39m(mapped), index\u001b[38;5;241m=\u001b[39mobj\u001b[38;5;241m.\u001b[39mindex)\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/base.py:921\u001b[0m, in \u001b[0;36mIndexOpsMixin._map_values\u001b[0;34m(self, mapper, na_action, convert)\u001b[0m\n\u001b[1;32m    918\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(arr, ExtensionArray):\n\u001b[1;32m    919\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m arr\u001b[38;5;241m.\u001b[39mmap(mapper, na_action\u001b[38;5;241m=\u001b[39mna_action)\n\u001b[0;32m--> 921\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43malgorithms\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmap_array\u001b[49m\u001b[43m(\u001b[49m\u001b[43marr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmapper\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mna_action\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mna_action\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconvert\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconvert\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/algorithms.py:1743\u001b[0m, in \u001b[0;36mmap_array\u001b[0;34m(arr, mapper, na_action, convert)\u001b[0m\n\u001b[1;32m   1741\u001b[0m values \u001b[38;5;241m=\u001b[39m arr\u001b[38;5;241m.\u001b[39mastype(\u001b[38;5;28mobject\u001b[39m, copy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m   1742\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m na_action \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 1743\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mlib\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmap_infer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmapper\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconvert\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconvert\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1744\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1745\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m lib\u001b[38;5;241m.\u001b[39mmap_infer_mask(\n\u001b[1;32m   1746\u001b[0m         values, mapper, mask\u001b[38;5;241m=\u001b[39misna(values)\u001b[38;5;241m.\u001b[39mview(np\u001b[38;5;241m.\u001b[39muint8), convert\u001b[38;5;241m=\u001b[39mconvert\n\u001b[1;32m   1747\u001b[0m     )\n",
      "File \u001b[0;32mlib.pyx:2972\u001b[0m, in \u001b[0;36mpandas._libs.lib.map_infer\u001b[0;34m()\u001b[0m\n",
      "Cell \u001b[0;32mIn[19], line 3\u001b[0m, in \u001b[0;36mget_bioproject\u001b[0;34m(bioproj_id)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_bioproject\u001b[39m(bioproj_id):\n\u001b[1;32m      2\u001b[0m     api_url \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhttps://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=bioproject&id=\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mbioproj_id\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m----> 3\u001b[0m     res \u001b[38;5;241m=\u001b[39m \u001b[43mrequests\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43mapi_url\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      4\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m res\u001b[38;5;241m.\u001b[39mtext\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/requests/api.py:76\u001b[0m, in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     65\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[1;32m     66\u001b[0m \n\u001b[1;32m     67\u001b[0m \u001b[38;5;124;03m:param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     72\u001b[0m \u001b[38;5;124;03m:rtype: requests.Response\u001b[39;00m\n\u001b[1;32m     73\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m     75\u001b[0m kwargs\u001b[38;5;241m.\u001b[39msetdefault(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m---> 76\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mget\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/requests/api.py:61\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;66;03m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;66;03m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     59\u001b[0m \u001b[38;5;66;03m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     60\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m sessions\u001b[38;5;241m.\u001b[39mSession() \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[0;32m---> 61\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msession\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/requests/sessions.py:544\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    539\u001b[0m send_kwargs \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    540\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m'\u001b[39m: timeout,\n\u001b[1;32m    541\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m'\u001b[39m: allow_redirects,\n\u001b[1;32m    542\u001b[0m }\n\u001b[1;32m    543\u001b[0m send_kwargs\u001b[38;5;241m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 544\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprep\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43msend_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    546\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/requests/sessions.py:657\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    654\u001b[0m start \u001b[38;5;241m=\u001b[39m preferred_clock()\n\u001b[1;32m    656\u001b[0m \u001b[38;5;66;03m# Send the request\u001b[39;00m\n\u001b[0;32m--> 657\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43madapter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    659\u001b[0m \u001b[38;5;66;03m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    660\u001b[0m elapsed \u001b[38;5;241m=\u001b[39m preferred_clock() \u001b[38;5;241m-\u001b[39m start\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/requests/adapters.py:439\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    437\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    438\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m chunked:\n\u001b[0;32m--> 439\u001b[0m         resp \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    440\u001b[0m \u001b[43m            \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    441\u001b[0m \u001b[43m            \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    442\u001b[0m \u001b[43m            \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    443\u001b[0m \u001b[43m            \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    444\u001b[0m \u001b[43m            \u001b[49m\u001b[43mredirect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    445\u001b[0m \u001b[43m            \u001b[49m\u001b[43massert_same_host\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    446\u001b[0m \u001b[43m            \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    447\u001b[0m \u001b[43m            \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    448\u001b[0m \u001b[43m            \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax_retries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    449\u001b[0m \u001b[43m            \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\n\u001b[1;32m    450\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    452\u001b[0m     \u001b[38;5;66;03m# Send the request.\u001b[39;00m\n\u001b[1;32m    453\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    454\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(conn, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mproxy_pool\u001b[39m\u001b[38;5;124m'\u001b[39m):\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/urllib3/connectionpool.py:700\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    697\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_prepare_proxy(conn)\n\u001b[1;32m    699\u001b[0m \u001b[38;5;66;03m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[0;32m--> 700\u001b[0m httplib_response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    701\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    702\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    703\u001b[0m \u001b[43m    \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    704\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout_obj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    705\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    706\u001b[0m \u001b[43m    \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    707\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    708\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    710\u001b[0m \u001b[38;5;66;03m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[1;32m    711\u001b[0m \u001b[38;5;66;03m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[1;32m    712\u001b[0m \u001b[38;5;66;03m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[1;32m    713\u001b[0m \u001b[38;5;66;03m# mess.\u001b[39;00m\n\u001b[1;32m    714\u001b[0m response_conn \u001b[38;5;241m=\u001b[39m conn \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m release_conn \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/urllib3/connectionpool.py:383\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    381\u001b[0m \u001b[38;5;66;03m# Trigger any extra validation we need to do.\u001b[39;00m\n\u001b[1;32m    382\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 383\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_conn\u001b[49m\u001b[43m(\u001b[49m\u001b[43mconn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    384\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (SocketTimeout, BaseSSLError) \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    385\u001b[0m     \u001b[38;5;66;03m# Py2 raises this as a BaseSSLError, Py3 raises it as socket timeout.\u001b[39;00m\n\u001b[1;32m    386\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_raise_timeout(err\u001b[38;5;241m=\u001b[39me, url\u001b[38;5;241m=\u001b[39murl, timeout_value\u001b[38;5;241m=\u001b[39mconn\u001b[38;5;241m.\u001b[39mtimeout)\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/urllib3/connectionpool.py:1017\u001b[0m, in \u001b[0;36mHTTPSConnectionPool._validate_conn\u001b[0;34m(self, conn)\u001b[0m\n\u001b[1;32m   1015\u001b[0m \u001b[38;5;66;03m# Force connect early to allow us to validate the connection.\u001b[39;00m\n\u001b[1;32m   1016\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(conn, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msock\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m):  \u001b[38;5;66;03m# AppEngine might not have  `.sock`\u001b[39;00m\n\u001b[0;32m-> 1017\u001b[0m     \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1019\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m conn\u001b[38;5;241m.\u001b[39mis_verified:\n\u001b[1;32m   1020\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m   1021\u001b[0m         (\n\u001b[1;32m   1022\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUnverified HTTPS request is being made to host \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1027\u001b[0m         InsecureRequestWarning,\n\u001b[1;32m   1028\u001b[0m     )\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/urllib3/connection.py:411\u001b[0m, in \u001b[0;36mHTTPSConnection.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    402\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (\n\u001b[1;32m    403\u001b[0m     \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mca_certs\n\u001b[1;32m    404\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mca_cert_dir\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    407\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(context, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mload_default_certs\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    408\u001b[0m ):\n\u001b[1;32m    409\u001b[0m     context\u001b[38;5;241m.\u001b[39mload_default_certs()\n\u001b[0;32m--> 411\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock \u001b[38;5;241m=\u001b[39m \u001b[43mssl_wrap_socket\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    412\u001b[0m \u001b[43m    \u001b[49m\u001b[43msock\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    413\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkeyfile\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mkey_file\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    414\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcertfile\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcert_file\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    415\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkey_password\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mkey_password\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    416\u001b[0m \u001b[43m    \u001b[49m\u001b[43mca_certs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mca_certs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    417\u001b[0m \u001b[43m    \u001b[49m\u001b[43mca_cert_dir\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mca_cert_dir\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    418\u001b[0m \u001b[43m    \u001b[49m\u001b[43mca_cert_data\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mca_cert_data\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    419\u001b[0m \u001b[43m    \u001b[49m\u001b[43mserver_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_hostname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    420\u001b[0m \u001b[43m    \u001b[49m\u001b[43mssl_context\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcontext\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    421\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtls_in_tls\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtls_in_tls\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    422\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    424\u001b[0m \u001b[38;5;66;03m# If we're using all defaults and the connection\u001b[39;00m\n\u001b[1;32m    425\u001b[0m \u001b[38;5;66;03m# is TLSv1 or TLSv1.1 we throw a DeprecationWarning\u001b[39;00m\n\u001b[1;32m    426\u001b[0m \u001b[38;5;66;03m# for the host.\u001b[39;00m\n\u001b[1;32m    427\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (\n\u001b[1;32m    428\u001b[0m     default_ssl_context\n\u001b[1;32m    429\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mssl_version \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    430\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mversion\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    431\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock\u001b[38;5;241m.\u001b[39mversion() \u001b[38;5;129;01min\u001b[39;00m {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTLSv1\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTLSv1.1\u001b[39m\u001b[38;5;124m\"\u001b[39m}\n\u001b[1;32m    432\u001b[0m ):\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/urllib3/util/ssl_.py:449\u001b[0m, in \u001b[0;36mssl_wrap_socket\u001b[0;34m(sock, keyfile, certfile, cert_reqs, ca_certs, server_hostname, ssl_version, ciphers, ssl_context, ca_cert_dir, key_password, ca_cert_data, tls_in_tls)\u001b[0m\n\u001b[1;32m    437\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m    438\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAn HTTPS request has been made, but the SNI (Server Name \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    439\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mIndication) extension to TLS is not available on this platform. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    445\u001b[0m         SNIMissingWarning,\n\u001b[1;32m    446\u001b[0m     )\n\u001b[1;32m    448\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m send_sni:\n\u001b[0;32m--> 449\u001b[0m     ssl_sock \u001b[38;5;241m=\u001b[39m \u001b[43m_ssl_wrap_socket_impl\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    450\u001b[0m \u001b[43m        \u001b[49m\u001b[43msock\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcontext\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtls_in_tls\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mserver_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_hostname\u001b[49m\n\u001b[1;32m    451\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    452\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    453\u001b[0m     ssl_sock \u001b[38;5;241m=\u001b[39m _ssl_wrap_socket_impl(sock, context, tls_in_tls)\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/urllib3/util/ssl_.py:493\u001b[0m, in \u001b[0;36m_ssl_wrap_socket_impl\u001b[0;34m(sock, ssl_context, tls_in_tls, server_hostname)\u001b[0m\n\u001b[1;32m    490\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m SSLTransport(sock, ssl_context, server_hostname)\n\u001b[1;32m    492\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m server_hostname:\n\u001b[0;32m--> 493\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mssl_context\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwrap_socket\u001b[49m\u001b[43m(\u001b[49m\u001b[43msock\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mserver_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_hostname\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    494\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    495\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m ssl_context\u001b[38;5;241m.\u001b[39mwrap_socket(sock)\n",
      "File \u001b[0;32m/usr/lib/python3.10/ssl.py:513\u001b[0m, in \u001b[0;36mSSLContext.wrap_socket\u001b[0;34m(self, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname, session)\u001b[0m\n\u001b[1;32m    507\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwrap_socket\u001b[39m(\u001b[38;5;28mself\u001b[39m, sock, server_side\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    508\u001b[0m                 do_handshake_on_connect\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m    509\u001b[0m                 suppress_ragged_eofs\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m    510\u001b[0m                 server_hostname\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, session\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    511\u001b[0m     \u001b[38;5;66;03m# SSLSocket class handles server_hostname encoding before it calls\u001b[39;00m\n\u001b[1;32m    512\u001b[0m     \u001b[38;5;66;03m# ctx._wrap_socket()\u001b[39;00m\n\u001b[0;32m--> 513\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msslsocket_class\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_create\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    514\u001b[0m \u001b[43m        \u001b[49m\u001b[43msock\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msock\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    515\u001b[0m \u001b[43m        \u001b[49m\u001b[43mserver_side\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_side\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    516\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdo_handshake_on_connect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdo_handshake_on_connect\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    517\u001b[0m \u001b[43m        \u001b[49m\u001b[43msuppress_ragged_eofs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msuppress_ragged_eofs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    518\u001b[0m \u001b[43m        \u001b[49m\u001b[43mserver_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_hostname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    519\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcontext\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    520\u001b[0m \u001b[43m        \u001b[49m\u001b[43msession\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msession\u001b[49m\n\u001b[1;32m    521\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/lib/python3.10/ssl.py:1100\u001b[0m, in \u001b[0;36mSSLSocket._create\u001b[0;34m(cls, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname, context, session)\u001b[0m\n\u001b[1;32m   1097\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0.0\u001b[39m:\n\u001b[1;32m   1098\u001b[0m             \u001b[38;5;66;03m# non-blocking\u001b[39;00m\n\u001b[1;32m   1099\u001b[0m             \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdo_handshake_on_connect should not be specified for non-blocking sockets\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m-> 1100\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdo_handshake\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1101\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (\u001b[38;5;167;01mOSError\u001b[39;00m, \u001b[38;5;167;01mValueError\u001b[39;00m):\n\u001b[1;32m   1102\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[0;32m/usr/lib/python3.10/ssl.py:1371\u001b[0m, in \u001b[0;36mSSLSocket.do_handshake\u001b[0;34m(self, block)\u001b[0m\n\u001b[1;32m   1369\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0.0\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m block:\n\u001b[1;32m   1370\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msettimeout(\u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[0;32m-> 1371\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sslobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdo_handshake\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1372\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m   1373\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msettimeout(timeout)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "projects['xml'] = projects['bioproject'].apply(get_bioproject)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "projects"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
